From 5544488dd797248ed8fcf648e9f7acc551774f30 Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesass.com>
Date: Wed, 23 Mar 2022 13:27:38 +0100
Subject: [PATCH 07/11] Remove debug messages

---
 drivers/pinctrl/renesas/pfc.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/drivers/pinctrl/renesas/pfc.c b/drivers/pinctrl/renesas/pfc.c
index b2066d5487..0e56b8ab3b 100644
--- a/drivers/pinctrl/renesas/pfc.c
+++ b/drivers/pinctrl/renesas/pfc.c
@@ -673,8 +673,6 @@ static int sh_pfc_pinconf_set_drive_strength(struct sh_pfc *pfc,
 	if (!reg)
 		return -EINVAL;
 
-	printf("Setting drive strength for PIN %d to %dmA\n", pin, strength);
-
 	step = size == 2 ? 6 : 3;
 
 	if (strength < step || strength > 24)
@@ -685,8 +683,6 @@ static int sh_pfc_pinconf_set_drive_strength(struct sh_pfc *pfc,
 	 */
 	strength = strength / step - 1;
 
-	printf("Strength reg val is %d\n", strength);
-
 	val = sh_pfc_read_raw_reg(reg, 32);
 	val &= ~GENMASK(offset + 4 - 1, offset);
 	val |= strength << offset;
-- 
2.17.1

