From f6a5d932d09b1356c651b28e263369c6886f0cb2 Mon Sep 17 00:00:00 2001
From: Michael Dege <michael.dege@renesas.com>
Date: Mon, 4 Jul 2022 10:44:55 +0200
Subject: [PATCH 08/11] DTS: VC4: add VCCQ node for controlling MMC IO voltages

---
 arch/arm/dts/r8a779f0-vc4.dts | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/arch/arm/dts/r8a779f0-vc4.dts b/arch/arm/dts/r8a779f0-vc4.dts
index 2bbac4fea5..c7a4d28183 100644
--- a/arch/arm/dts/r8a779f0-vc4.dts
+++ b/arch/arm/dts/r8a779f0-vc4.dts
@@ -58,6 +58,18 @@
 		gpio = <&gpio1 24 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
 	};
+
+	vccq_sdhi: regulator-vccq-sdhi {
+		compatible = "regulator-gpio";
+
+		regulator-name = "SDHI VccQ";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+
+		gpios = <&gpio2 16 GPIO_ACTIVE_HIGH>;
+		gpios-states = <1>;
+		states = <3300000 1>, <1800000 0>;
+	};
 };
 
 &extal_clk {
@@ -115,7 +127,7 @@
 	pinctrl-names = "default", "state_uhs";
 
 	vmmc-supply = <&vcc_sdhi>;
-	vqmmc-supply = <&vcc_sdhi>;
+	vqmmc-supply = <&vccq_sdhi>;
 	cd-gpios = <&gpio1 23 GPIO_ACTIVE_LOW>;
 	bus-width = <4>;
 };
-- 
2.17.1

