From b689e82e1fcccd2406bda6ec310f1426536cec1e Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesass.com>
Date: Mon, 21 Mar 2022 09:30:22 +0100
Subject: [PATCH 040/151] net: phy: m88q211x sanitize return values

---
 drivers/net/phy/marvell_88q211x.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/phy/marvell_88q211x.c b/drivers/net/phy/marvell_88q211x.c
index 9874d4004a2e..b931826d25ba 100644
--- a/drivers/net/phy/marvell_88q211x.c
+++ b/drivers/net/phy/marvell_88q211x.c
@@ -840,7 +840,7 @@ static int phy_init(struct phy_device *phydev, uint16_t flags)
 		init_100mbit(phydev);
 	}
 	else {
-		return MRVL_88Q2112_INIT_ERROR_ARGS;
+		return -MRVL_88Q2112_INIT_ERROR_ARGS;
 	}
 
 	// 3) Timeout until link-up and compatibility monitor
@@ -849,7 +849,8 @@ static int phy_init(struct phy_device *phydev, uint16_t flags)
 		if (check_link(phydev)) {
 			// link up, exit with success
 			printk("%s:%d INIT_DONE\n", __FUNCTION__, __LINE__);
-			return MRVL_88Q2112_INIT_DONE;
+			//return MRVL_88Q2112_INIT_DONE;
+			return 0;
 		}
 		else {
 			// 1000 BASE-T1: During training stage, PHYs can exchange PHY Revision information.
@@ -876,7 +877,8 @@ static int phy_init(struct phy_device *phydev, uint16_t flags)
 			if (check_link(phydev)) {
 				// link up, exit with success
 				printk("%s:%d INIT_DONE\n", __FUNCTION__, __LINE__);
-				return MRVL_88Q2112_INIT_DONE;
+				//return MRVL_88Q2112_INIT_DONE;
+				return 0;
 			}
 			// no else statement
 		}
@@ -885,7 +887,7 @@ static int phy_init(struct phy_device *phydev, uint16_t flags)
 	// Init failed - no link within expected time.
 	// 1000 BASE-T1: a loop could be implemented
 	// to toggle back into MRVL_88Q2112_MODE_DEFAULT and try again.
-	return MRVL_88Q2112_INIT_ERROR_FAILED;
+	return -MRVL_88Q2112_INIT_ERROR_FAILED;
 }
 
 static struct phy_driver mv88q211x_drivers[] = {
-- 
2.17.1

