From 8b2117a87b9a35cd3af0f19ec685128696e567b5 Mon Sep 17 00:00:00 2001
From: "Christian@p620" <Christian@p620>
Date: Wed, 24 Aug 2022 14:48:35 +0200
Subject: [PATCH 088/151] Activate SPI also in V1 device tree

---
 .../renesas/r8a779f0-vc4-ctrl-domainV1.dts    | 75 +++++++++++++++++++
 1 file changed, 75 insertions(+)

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domainV1.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domainV1.dts
index 10e66ece4a90..4a0b85eab83b 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domainV1.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domainV1.dts
@@ -166,6 +166,66 @@
 	};
 };
 
+&mspi0 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi0_pins>;
+	pinctrl-names = "default";
+
+	t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <10000000>;
+
+		interrupt-parent = <&gpio6>;
+		interrupts = <18 IRQ_TYPE_LEVEL_LOW>;
+
+		plca-enable = /bits/ 8 <0>;
+		tx-cut-through-mode = /bits/ 8 <1>;
+	};
+};
+
+&mspi1 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi1_pins>;
+	pinctrl-names = "default";
+
+	t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <10000000>;
+
+		interrupt-parent = <&gpio6>;
+		interrupts = <19 IRQ_TYPE_LEVEL_LOW>;
+
+		plca-enable = /bits/ 8 <0>;
+		tx-cut-through-mode = /bits/ 8 <1>;
+	};
+};
+
+&mspi4 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi4_pins>;
+	pinctrl-names = "default";
+
+	t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <10000000>;
+
+		interrupt-parent = <&gpio6>;
+		interrupts = <20 IRQ_TYPE_LEVEL_LOW>;
+
+		plca-enable = /bits/ 8 <0>;
+		tx-cut-through-mode = /bits/ 8 <1>;
+	};
+};
+
 &pfc {
 	canfd0_pins: canfd0 {
 		channel0 {
@@ -241,4 +301,19 @@
 		groups = "etnb0_rmii", "etnb0_mdio", "etnb0_link";
 		function = "etnb0";
 	};
+
+	mspi0_pins: mpsi0 {
+		groups = "mspi0_sck", "mspi0_din", "mspi0_dout", "mspi0_cs0";
+		function = "mspi0";
+	};
+
+	mspi1_pins: mpsi1 {
+		groups = "mspi1_sck", "mspi1_din", "mspi1_dout", "mspi1_cs0";
+		function = "mspi1";
+	};
+
+	mspi4_pins: mpsi4 {
+		groups = "mspi4_sck", "mspi4_din", "mspi4_dout", "mspi4_cs0";
+		function = "mspi4";
+	};
 };
-- 
2.17.1

