From 9939f3d288da2d36db0113a26466e41435e9aefe Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesass.com>
Date: Thu, 8 Sep 2022 11:28:29 +0200
Subject: [PATCH 096/151] Add ethtool support and IOCTLs for HW time stamping

---
 .../ethernet/renesas/rswitch2/rswitch2_eth.c  | 121 +++++++++++++++++-
 1 file changed, 116 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
index 4a3f68eb1afb..96e74ca5f977 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
@@ -28,7 +28,7 @@
 #include "rswitch2_gwca.h"
 #include "rswitch2_rmac.h"
 #include "rswitch2_serdes.h"
-
+#include "../rtsn_ptp.h"
 
 static int rswitch2_gwca_set_state(struct rswitch2_drv *rsw2, enum gwmc_op state)
 {
@@ -1463,22 +1463,109 @@ static void rswitch2_eth_tx_timeout(struct net_device *ndev, unsigned int txqueu
 {
 }
 
+static int rswitch2_hwstamp_get(struct net_device *ndev, struct ifreq *req)
+{
+	struct rswitch2_drv *rsw2;
+	struct rswitch2_eth_port *eth_port;
+	struct rswitch2_internal_port *intern_port;
+	struct rswitch2_physical_port *phy_port;
+	struct rtsn_ptp_private *ptp_priv;
+	struct hwtstamp_config config;
+	eth_port = netdev_priv(ndev);
+	intern_port = eth_port->intern_port;
+	phy_port = eth_port->phy_port;
+	if(intern_port) {
+		printk("HW get intern port!?\n");
+		return -EINVAL;
+	}
+	rsw2 = eth_port->rsw2;
+	ptp_priv = rsw2->ptp_drv;
+	config.flags = 0;
+	config.tx_type = ptp_priv->tstamp_tx_ctrl ? HWTSTAMP_TX_ON :
+						    HWTSTAMP_TX_OFF;
+	switch (ptp_priv->tstamp_rx_ctrl & RTSN_RXTSTAMP_TYPE) {
+	case RTSN_RXTSTAMP_TYPE_V2_L2_EVENT:
+		config.rx_filter = HWTSTAMP_FILTER_PTP_V2_L2_EVENT;
+		break;
+	case RTSN_RXTSTAMP_TYPE_ALL:
+		config.rx_filter = HWTSTAMP_FILTER_ALL;
+		break;
+	default:
+		config.rx_filter = HWTSTAMP_FILTER_NONE;
+		break;
+	}
+	return copy_to_user(req->ifr_data, &config, sizeof(config)) ? -EFAULT : 0;
+}
+static int rswitch2_hwstamp_set(struct net_device *ndev, struct ifreq *req)
+{
+	struct rswitch2_drv *rsw2;
+	struct rswitch2_eth_port *eth_port;
+	struct rswitch2_internal_port *intern_port;
+	struct rswitch2_physical_port *phy_port;
+	struct rtsn_ptp_private *ptp_priv;
+	struct hwtstamp_config config;
+	u32 tstamp_rx_ctrl = RTSN_RXTSTAMP_ENABLED;
+	u32 tstamp_tx_ctrl;
+	eth_port = netdev_priv(ndev);
+	intern_port = eth_port->intern_port;
+	phy_port = eth_port->phy_port;
+	if(intern_port) {
+		printk("HW set intern port!?\n");
+		return -EINVAL;
+	}
+	rsw2 = eth_port->rsw2;
+	ptp_priv = rsw2->ptp_drv;
+	if (copy_from_user(&config, req->ifr_data, sizeof(config)))
+		return -EFAULT;
+	if (config.flags)
+		return -EINVAL;
+	switch (config.tx_type) {
+	case HWTSTAMP_TX_OFF:
+		tstamp_tx_ctrl = 0;
+		break;
+	case HWTSTAMP_TX_ON:
+		tstamp_tx_ctrl = RTSN_TXTSTAMP_ENABLED;
+		break;
+	default:
+		return -ERANGE;
+	}
+	switch (config.rx_filter) {
+	case HWTSTAMP_FILTER_NONE:
+		tstamp_rx_ctrl = 0;
+		break;
+	case HWTSTAMP_FILTER_PTP_V2_L2_EVENT:
+		tstamp_rx_ctrl |= RTSN_RXTSTAMP_TYPE_V2_L2_EVENT;
+		break;
+	default:
+		config.rx_filter = HWTSTAMP_FILTER_ALL;
+		tstamp_rx_ctrl |= RTSN_RXTSTAMP_TYPE_ALL;
+		break;
+	}
+	ptp_priv->tstamp_tx_ctrl = tstamp_tx_ctrl;
+	ptp_priv->tstamp_rx_ctrl = tstamp_rx_ctrl;
+	return copy_to_user(req->ifr_data, &config, sizeof(config)) ? -EFAULT : 0;
+}
 static int rswitch2_eth_do_ioctl(struct net_device *ndev, struct ifreq *req, int cmd)
 {
 	struct phy_device *phydev = ndev->phydev;
 
-	if (!netif_running(ndev))
+	if (!netif_running(ndev)) {
+		printk("netif not running\n");
 		return -EINVAL;
+	}
 
-	if (!phydev)
+	if (!phydev) {
+		printk("no phydev\n");
 		return -ENODEV;
-
+	}
 	switch (cmd) {
 	case SIOCGHWTSTAMP:
 		/* HW get timestamp */
+		return rswitch2_hwstamp_get(ndev, req);
 		break;
 	case SIOCSHWTSTAMP:
 		/* HW set timestamp */
+		return rswitch2_hwstamp_set(ndev, req);
 		break;
 	}
 
@@ -1890,6 +1977,29 @@ static int rswitch2_set_ringparam(struct net_device *ndev,
 	/* TODO */
 	return 0;
 }
+static int rswitch2_get_ts_info(struct net_device *ndev, struct ethtool_ts_info *info)
+{
+	struct rswitch2_eth_port *eth_port;
+	struct rswitch2_drv *rsw2;
+	eth_port = netdev_priv(ndev);
+	rsw2 = eth_port->rsw2;
+	if(eth_port->intern_port) {
+		info->so_timestamping = SOF_TIMESTAMPING_RX_SOFTWARE |
+					SOF_TIMESTAMPING_SOFTWARE;
+		info->phc_index = -1;
+		return 0;
+	}
+	info->phc_index = ptp_clock_index(rsw2->ptp_drv->clock);
+	info->so_timestamping = SOF_TIMESTAMPING_TX_SOFTWARE |
+				SOF_TIMESTAMPING_RX_SOFTWARE |
+				SOF_TIMESTAMPING_SOFTWARE |
+				SOF_TIMESTAMPING_TX_HARDWARE |
+				SOF_TIMESTAMPING_RX_HARDWARE |
+				SOF_TIMESTAMPING_RAW_HARDWARE;
+	info->tx_types = BIT(HWTSTAMP_TX_OFF) | BIT(HWTSTAMP_TX_ON);
+	info->rx_filters = BIT(HWTSTAMP_FILTER_NONE) | BIT(HWTSTAMP_FILTER_ALL);
+	return 0;
+}
 
 static const struct ethtool_ops rswitch2_ethtool_ops = {
 	.nway_reset			= phy_ethtool_nway_reset,
@@ -1901,7 +2011,7 @@ static const struct ethtool_ops rswitch2_ethtool_ops = {
 	.get_sset_count		= rswicth2_get_sset_count,
 	.get_ringparam		= rswitch2_get_ringparam,
 	.set_ringparam		= rswitch2_set_ringparam,
-	.get_ts_info		= NULL,
+	.get_ts_info		= rswitch2_get_ts_info,
 	.get_link_ksettings	= phy_ethtool_get_link_ksettings,
 	.set_link_ksettings	= phy_ethtool_set_link_ksettings,
 	.get_wol			= NULL,
@@ -2406,6 +2516,7 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 	port_ndev->min_mtu = 2048 - (ETH_HLEN + VLAN_HLEN + ETH_FCS_LEN);
 	port_ndev->max_mtu = ETH_MIN_MTU;
 	port_ndev->netdev_ops = &rswitch2_netdev_ops;
+	port_ndev->ethtool_ops = &rswitch2_ethtool_ops;
 
 	ret = rswitch2_mdio_init(port_ndev, port_num);
 	if (ret != 0) {
-- 
2.17.1

