From 6e3599e6926d9f1365d0e35c0da0d9c16e330d1c Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesass.com>
Date: Thu, 8 Dec 2022 16:30:48 +0100
Subject: [PATCH 129/151] Checkout for invalid PHY configuration

---
 .../ethernet/renesas/rswitch2/rswitch2_eth.c  | 26 ++++++++++++++-----
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
index aedfc0ea8dfe..7dfeb2e45977 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
@@ -2622,7 +2622,7 @@ static const struct ethtool_ops rswitch2_ethtool_ops = {
 };
 
 
-static void rswitch2_init_port_mac(struct net_device *port_ndev)
+static int rswitch2_init_port_mac(struct net_device *port_ndev)
 {
 	u32 reg_val;
 	int speed;
@@ -2661,7 +2661,7 @@ static void rswitch2_init_port_mac(struct net_device *port_ndev)
 			break;
 		default:
 			rsw2_err(MSG_GEN, "Unsupported MAC xMII format %s (%d) on port %d\n",phy_modes(phy_iface), phy_iface, eth_port->port_num-1);
-			//return -EINVAL;
+			return -EINVAL;
 	}
 	printk("%s uses %s at %d Mbps\n", port_ndev->name, phy_modes(phy_iface), speed);
 
@@ -2681,6 +2681,8 @@ static void rswitch2_init_port_mac(struct net_device *port_ndev)
 	reg_val |= MRAFC_MCENP;
 	reg_val |= MRAFC_UCENP;
 	iowrite32(reg_val, phy_port->rmac_base_addr + RSW2_RMAC_MRAFC);
+
+	return 0;
 }
 
 
@@ -2984,6 +2986,9 @@ static int rswitch2_get_phy_config(struct net_device *ndev)
 		else {
 			rsw2_info(MSG_GEN, "Got phy_iface mode: %s (%d)\n", phy_modes(phy_port->phy_iface), phy_port->phy_iface);
 		}
+	} else {
+		rsw2_err(MSG_GEN, "Failed to get port nodeof_get_phy_mode failed\n");
+		return -ENODEV;
 	}
 	of_node_put(port_node);
 
@@ -3075,7 +3080,11 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 	/* MAC related setup */
 	rswitch2_init_mac_addr(port_ndev);
 	rswitch2_port_set_mac_addr(port_ndev);
-	rswitch2_init_port_mac(port_ndev);
+	ret = rswitch2_init_port_mac(port_ndev);
+	if (ret < 0) {
+		netdev_err(port_ndev, "Could not init port MAC: %d\n", ret);
+		goto cleanup_phy_port;
+	}
 
 	/* Change to OPERATION Mode */
 	ret = rswitch2_emac_set_state(port_ndev, emac_operation);
@@ -3462,18 +3471,17 @@ int rswitch2_eth_init(struct rswitch2_drv *rsw2)
 				rsw2_err(MSG_GEN, "RX ring %d format for '%s' failed\n", cur_rx_q, ndev->name);
 				goto port_init_err;
 			}
-
-
 		}
 	}
 
 	for (cur_port = 0; cur_port < rsw2->num_of_tsn_ports; cur_port++) {
 		struct net_device *ndev;
 
-		rswitch2_init_physical_port(rsw2, cur_port);
+
+		ret = rswitch2_init_physical_port(rsw2, cur_port);
 		if (ret < 0) {
 			rsw2_err(MSG_GEN, "Failed to initialize port %d: %d\n", cur_port, ret);
-			goto port_init_err;
+			continue;
 		}
 		rsw2_info(MSG_GEN, "rswitch2_init_physical_port(%d) done\n", cur_port);
 		ndev = rsw2->ports[cur_port + rsw2->num_of_cpu_ports]->ndev;
@@ -3522,6 +3530,10 @@ int rswitch2_eth_init(struct rswitch2_drv *rsw2)
 
 	for (cur_port = 0; cur_port < rsw2->num_of_tsn_ports + rsw2->num_of_cpu_ports; cur_port++) {
 		struct net_device *ndev;
+
+		if(!rsw2->ports[cur_port])
+			continue;
+
 		ndev = rsw2->ports[cur_port ]->ndev;
 		ret = register_netdev(ndev);
 		if (ret < 0) {
-- 
2.17.1

