From 7c40008becc59df85401d645761c432dfc2ef5e8 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Thu, 5 May 2022 16:36:31 +0300
Subject: [PATCH 052/157] arm64: dts: renesas: vc4: add ethb0 (ether-avb)
 device

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi b/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
index 4fed6d3bed57..1d4ca06280c3 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
@@ -173,6 +173,25 @@ channel7 {
 		};
 	};
 
+	etnb0: ethernet@df0a2000 {
+		compatible = "renesas,etheravb-r8a779f0";
+		reg = <0 0xdf0a2000 0 0x800>;
+		clocks = <&cpg CPG_CD_MOD 200>;
+		clock-names = "fck";
+		interrupt-parent = <&intswcd>;
+		interrupts = <R8A779F0_INTSWCD_INTETNB0DATA>,
+			     <R8A779F0_INTSWCD_INTETNB0ERR>,
+			     <R8A779F0_INTSWCD_INTETNB0MNG>,
+			     <R8A779F0_INTSWCD_INTETNB0MAC>;
+		interrupt-names = "data", "error", "management", "mac";
+		/* power domain is responsible for enabling/disabling clock
+		   in dynamic pm operations */
+		power-domains = <&sysc R8A779F0_PD_ALWAYS_ON>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+	};
+
 	gpio4: gpio@dfd90180 {
 		compatible = "renesas,gpio-r8a779f0";
 		reg = <0 0xdfd90180 0 0x54>;
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
index af64566d4f91..301a10a96723 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
@@ -146,6 +146,39 @@ channel7 {
 	};
 };
 
+&etnb0 {
+	/* keep it disabled until driver support is implemented */
+	status = "okay";
+
+	pinctrl-0 = <&etnb0_pins>;
+	pinctrl-names = "default";
+
+	/* cpu-phys-addr dma-addr size */
+	bounce-buffer = <0xde000000 0xfe000000 0x00080000>; /* cluster0 ram */
+
+	phy-mode = "rmii";
+	phy-handle = <&etnb0_100base_tx_phy>;
+
+	etnb0_100base_tx_phy: ethernet-phy@1 {
+		reg = <1>;
+		status = "okay";
+	};
+
+	etnb0_100base_t1_phy: ethernet-phy@2 {
+		reg = <2>;
+		status = "disabled";
+	};
+};
+
+&gpio5 {
+	etnb0-mux {
+		gpio-hog;
+		gpios = <17 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "ETNB0-MUX";
+	};
+};
+
 &gpio6 {
 	sel-can-sent {
 		gpio-hog;
@@ -232,4 +265,9 @@ channel15 {
 			function = "canfd15";
 		};
 	};
+
+	etnb0_pins: etnb0 {
+		groups = "etnb0_rmii", "etnb0_mdio", "etnb0_link";
+		function = "etnb0";
+	};
 };
-- 
2.25.1

