From bf17a3b6153e4c8549f1fcffcfa4e3bceb33b103 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Thu, 25 May 2023 10:46:48 +0600
Subject: [PATCH 62/85] arm64: dts: renesas: vc4: add LAN8650 devices

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 .../dts/renesas/r8a779f0-vc4-ctrl-domain.dts  | 127 ++++++++++++++++++
 1 file changed, 127 insertions(+)

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
index f5916206af40..a954dca94055 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
@@ -184,6 +184,103 @@ sel-can-flxray {
 	};
 };
 
+&intswcd {
+	quirks {
+		intp35 {
+			interrupt = <R8A779F0_INTSWCD_INTP35>;
+			force-level-low;
+		};
+
+		intp34 {
+			interrupt = <R8A779F0_INTSWCD_INTP34>;
+			force-level-low;
+		};
+
+		intp33 {
+			interrupt = <R8A779F0_INTSWCD_INTP33>;
+			force-level-low;
+		};
+	};
+};
+
+&mspi0 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi0_pins>;
+	pinctrl-names = "default";
+
+	t1s0: t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <20000000>;
+
+		pinctrl-0 = <&t1s0_pins>;
+		pinctrl-names = "default";
+
+		interrupt-parent = <&intswcd>;
+		interrupts = <R8A779F0_INTSWCD_INTP35>;
+
+		tx-cut-through-mode = /bits/ 8 <0>;
+
+		plca-enable = /bits/ 8 <0>;
+		plca-node-id = /bits/ 8 <0>;	/* range: 0 to 254 */
+		plca-node-count = /bits/ 8 <8>; /* range: 1 to 255 */
+	};
+};
+
+&mspi1 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi1_pins>;
+	pinctrl-names = "default";
+
+	t1s1: t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <20000000>;
+
+		pinctrl-0 = <&t1s1_pins>;
+		pinctrl-names = "default";
+
+		interrupt-parent = <&intswcd>;
+		interrupts = <R8A779F0_INTSWCD_INTP34>;
+
+		tx-cut-through-mode = /bits/ 8 <0>;
+
+		plca-enable = /bits/ 8 <0>;
+		plca-node-id = /bits/ 8 <0>;	/* range: 0 to 254 */
+		plca-node-count = /bits/ 8 <8>; /* range: 1 to 255 */
+	};
+};
+
+&mspi4 {
+	status = "okay";
+
+	pinctrl-0 = <&mspi4_pins>;
+	pinctrl-names = "default";
+
+	t1s2: t1s@0 {
+		compatible = "microchip,lan865x";
+		reg = <0>;
+
+		spi-max-frequency = <20000000>;
+
+		pinctrl-0 = <&t1s2_pins>;
+		pinctrl-names = "default";
+
+		interrupt-parent = <&intswcd>;
+		interrupts = <R8A779F0_INTSWCD_INTP33>;
+
+		tx-cut-through-mode = /bits/ 8 <0>;
+
+		plca-enable = /bits/ 8 <0>;
+		plca-node-id = /bits/ 8 <0>;	/* range: 0 to 254 */
+		plca-node-count = /bits/ 8 <8>; /* range: 1 to 255 */
+	};
+};
+
 &pfc {
 	canfd0_pins: canfd0 {
 		channel0 {
@@ -259,4 +356,34 @@ etnb0_pins: etnb0 {
 		groups = "etnb0_rmii", "etnb0_mdio", "etnb0_link";
 		function = "etnb0";
 	};
+
+	mspi0_pins: mpsi0 {
+		groups = "mspi0_sck", "mspi0_din", "mspi0_dout", "mspi0_cs0";
+		function = "mspi0";
+	};
+
+	mspi1_pins: mpsi1 {
+		groups = "mspi1_sck", "mspi1_din", "mspi1_dout", "mspi1_cs0";
+		function = "mspi1";
+	};
+
+	mspi4_pins: mpsi4 {
+		groups = "mspi4_sck", "mspi4_din", "mspi4_dout", "mspi4_cs0";
+		function = "mspi4";
+	};
+
+	t1s0_pins: t1s0 {
+		groups = "intp35";
+		function = "intp";
+	};
+
+	t1s1_pins: t1s1 {
+		groups = "intp34";
+		function = "intp";
+	};
+
+	t1s2_pins: t1s2 {
+		groups = "intp33";
+		function = "intp";
+	};
 };
-- 
2.39.2

