From a65238cf3ef578cc2abe272a39b88f3566109603 Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesas.com>
Date: Wed, 14 Jun 2023 15:14:39 +0200
Subject: [PATCH 64/85] net: phy: swphy: Add support for 2500MBit speed

Signed-off-by: Dennis Ostermann <dennis.ostermann@renesas.com>
---
 drivers/net/phy/swphy.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/phy/swphy.c b/drivers/net/phy/swphy.c
index 59f1ba4d49bc..a038e69a5ce7 100644
--- a/drivers/net/phy/swphy.c
+++ b/drivers/net/phy/swphy.c
@@ -29,6 +29,7 @@ enum {
 	SWMII_SPEED_10 = 0,
 	SWMII_SPEED_100,
 	SWMII_SPEED_1000,
+	SWMII_SPEED_2500,
 	SWMII_DUPLEX_HALF = 0,
 	SWMII_DUPLEX_FULL,
 };
@@ -51,6 +52,11 @@ static const struct swmii_regs speed[] = {
 		.lpagb = LPA_1000FULL | LPA_1000HALF,
 		.estat = ESTATUS_1000_TFULL | ESTATUS_1000_THALF,
 	},
+	[SWMII_SPEED_2500] = {
+		.bmsr  = BMSR_ESTATEN,
+		.lpagb = MDIO_USXGMII_2500FULL | MDIO_USXGMII_2500HALF,
+		.estat = MDIO_AN_10GBT_STAT_LP2_5G,
+	},
 };
 
 static const struct swmii_regs duplex[] = {
@@ -71,6 +77,8 @@ static const struct swmii_regs duplex[] = {
 static int swphy_decode_speed(int speed)
 {
 	switch (speed) {
+	case 2500:
+		return SWMII_SPEED_2500;
 	case 1000:
 		return SWMII_SPEED_1000;
 	case 100:
-- 
2.39.2

