From b01239aeccc450af73bed38c200ed0979fa7b24f Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Tue, 31 Oct 2023 11:02:11 +0600
Subject: [PATCH 070/157] arm64: dts: renesas: vc4: add ufs device

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
index 42a78fe51fe9..b8d3c5d055ad 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
@@ -385,3 +385,11 @@ &scif_clk {
 
 	clock-frequency = <24000000>;
 };
+
+&ufs {
+	status = "okay";
+};
+
+&ufs30_clk {
+	clock-frequency = <38400000>;
+};
-- 
2.25.1

