From b438a59df963c7738cbc56ce8acc37f5b5d066d7 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 7 Oct 2024 11:37:55 +0500
Subject: [PATCH 145/157] net: ethernet: renesas: rswitch: add TOP register
 definitions

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/drivers/net/ethernet/renesas/rswitch.c b/drivers/net/ethernet/renesas/rswitch.c
index c5b824471110..a91c1bf4010e 100644
--- a/drivers/net/ethernet/renesas/rswitch.c
+++ b/drivers/net/ethernet/renesas/rswitch.c
@@ -78,6 +78,7 @@
 
 #define GWCA_IPV_NUM		0
 
+#define RSWITCH_TOP_OFFSET	0x00008000
 #define RSWITCH_COMA_OFFSET	0x00009000
 #define RSWITCH_ETHA_OFFSET	0x0000a000	/* with RMAC */
 #define RSWITCH_ETHA_SIZE	0x00002000	/* with RMAC */
@@ -86,6 +87,7 @@
 #define RSWITCH_GPTP_OFFSET    0x00018000
 
 #define FWRO	0
+#define TPRO	RSWITCH_TOP_OFFSET
 #define CARO	RSWITCH_COMA_OFFSET
 #if GWCA_INDEX == 1
 #define GWRO	RSWITCH_GWCA1_OFFSET
@@ -460,6 +462,23 @@ enum rswitch_reg {
 	FWSCR45		= FWRO + 0x7DB4,
 	FWSCR46		= FWRO + 0x7DB8,
 
+	TPEMIMC0	= TPRO + 0x0000,
+	TPEMIMC1	= TPRO + 0x0004,
+	TPEMIMC2	= TPRO + 0x0008,
+	TPEMIMC3	= TPRO + 0x000C,
+	TPEMIMC4	= TPRO + 0x0010,
+	TPEMIMC5	= TPRO + 0x0014,
+	TPEMIMC60	= TPRO + 0x0080,
+	TPEMIMC70	= TPRO + 0x0100,
+	TSIM		= TPRO + 0x0700,
+	TFIM		= TPRO + 0x0704,
+	TCIM		= TPRO + 0x0708,
+	TGIM0		= TPRO + 0x0710,
+	TGIM1		= TPRO + 0x0714,
+	TEIM0		= TPRO + 0x0720,
+	TEIM1		= TPRO + 0x0724,
+	TEIM2		= TPRO + 0x0728,
+
 	RIPV		= CARO + 0x0000,
 	RRC		= CARO + 0x0004,
 	RCEC		= CARO + 0x0008,
-- 
2.25.1

