From 3b1f3cb96dcbf3ad834f420f47198f1afe07c5e9 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 14 Oct 2024 09:38:39 +0500
Subject: [PATCH 147/157] net: ethernet: renesas: rswitch: configure GWCA MAC
 address

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/drivers/net/ethernet/renesas/rswitch.c b/drivers/net/ethernet/renesas/rswitch.c
index a54d265e28e6..3c324b7d6d38 100644
--- a/drivers/net/ethernet/renesas/rswitch.c
+++ b/drivers/net/ethernet/renesas/rswitch.c
@@ -1049,6 +1049,8 @@ struct rswitch_gwca {
 	DECLARE_BITMAP(used, RSWITCH_MAX_NUM_CHAINS);
 
 	u32 external_irq[RSWITCH_NUM_IRQ_REGS];
+
+	u8 mac_addr[ETH_ALEN];
 };
 
 struct rswitch_route_monitor {
@@ -2648,6 +2650,16 @@ static int rswitch_gwca_axi_ram_reset(struct rswitch_private *priv)
 	return rswitch_gwca_poll(priv, GWARIRM, GWARIRM_ARR, GWARIRM_ARR);
 }
 
+static void rswitch_gwca_write_mac_address(struct rswitch_private *priv)
+{
+	u8 *mac = &priv->gwca.mac_addr[0];
+	u32 gwmac0 = (mac[0] << 8) | mac[1];
+	u32 gwmac1 = (mac[2] << 24) | (mac[3] << 16) | (mac[4] << 8) | mac[5];
+
+	rswitch_gwca_write(priv, GWMAC0, gwmac0);
+	rswitch_gwca_write(priv, GWMAC1, gwmac1);
+}
+
 static int rswitch_gwca_hw_init(struct rswitch_private *priv)
 {
 	int err;
@@ -2677,6 +2689,8 @@ static int rswitch_gwca_hw_init(struct rswitch_private *priv)
 	rswitch_gwca_write(priv, GWMDNC, (0xff << 8) | 0xff);
 	rswitch_gwca_write(priv, GWTPC0, GWTPC_PPPL(GWCA_IPV_NUM));
 
+	rswitch_gwca_write_mac_address(priv);
+
 	err = rswitch_gwca_change_mode(priv, GWMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
@@ -4625,6 +4639,8 @@ static int rswitch_probe(struct platform_device *pdev)
 			return ret;
 	}
 
+	eth_random_addr(priv->gwca.mac_addr);
+
 	INIT_LIST_HEAD(&priv->port_list);
 	ports_dn = of_get_child_by_name(priv->pdev->dev.of_node, "ports");
 	if (!ports_dn) {
-- 
2.25.1

