From 66936d3e15d341108352393eddb579be48832413 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 2 Sep 2024 09:23:40 +0500
Subject: [PATCH 148/157] net: ethernet: renesas: rswitch: initial vport
 implementation

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/drivers/net/ethernet/renesas/Kconfig b/drivers/net/ethernet/renesas/Kconfig
index 4f8e4c875790..9ba5e37f899c 100644
--- a/drivers/net/ethernet/renesas/Kconfig
+++ b/drivers/net/ethernet/renesas/Kconfig
@@ -62,6 +62,10 @@ config RENESAS_ETHER_SWITCH
 	  This driver supports the following SoCs:
 		- R8A779Fx.
 
+config RENESAS_ETHER_SWITCH_VPORT_LOCAL
+	tristate "Local rswitch virtual port"
+	depends on RENESAS_ETHER_SWITCH
+
 config RTSN
         tristate "Renesas Ethernet-TSN support"
         depends on ARCH_RENESAS || COMPILE_TEST
diff --git a/drivers/net/ethernet/renesas/Makefile b/drivers/net/ethernet/renesas/Makefile
index e12589f402e6..f34d12211477 100644
--- a/drivers/net/ethernet/renesas/Makefile
+++ b/drivers/net/ethernet/renesas/Makefile
@@ -13,3 +13,4 @@ obj-$(CONFIG_RTSN_PTP) += rtsn_ptp.o
 obj-$(CONFIG_RTSN) += rtsn.o
 obj-$(CONFIG_RENESAS_ETHER_SWITCH_LIB) += rswitch_lib.o
 obj-$(CONFIG_RENESAS_ETHER_SWITCH) += rswitch.o
+obj-$(CONFIG_RENESAS_ETHER_SWITCH_VPORT_LOCAL) += rswitch_vport_local.o
diff --git a/drivers/net/ethernet/renesas/rswitch.c b/drivers/net/ethernet/renesas/rswitch.c
index 3c324b7d6d38..e7d6e2bc033d 100644
--- a/drivers/net/ethernet/renesas/rswitch.c
+++ b/drivers/net/ethernet/renesas/rswitch.c
@@ -1111,6 +1111,30 @@ struct rswitch_device {
 	unsigned int l3_offload_enabled : 1;
 };
 
+struct rswitch_vport_backend {
+	struct rswitch_private *priv;
+	struct list_head list;
+	unsigned int id;
+
+	struct rswitch_vport_desc *rvpd;
+
+	/* backend chain ids */
+	int tx_chain_index;
+	int rx_chain_index;
+
+	/* backend rings */
+	struct rswitch_tx_ring *tx_ring;
+	struct rswitch_rx_ring *rx_ring;
+
+	/* backend netdev */
+	struct net_device *ndev;
+	struct napi_struct napi;
+
+	/* frontend chain ids */
+	int fe_tx_chain_index;
+	int fe_rx_chain_index;
+};
+
 struct rswitch_private {
 	struct platform_device *pdev;
 	void __iomem *addr;
@@ -1137,6 +1161,9 @@ struct rswitch_private {
 
 	u32 parallel_mode_saved_clocks;
 
+	struct list_head vport_list;
+	unsigned long vport_ids;
+
 	struct net_device *offload_brdev;
 
 	spinlock_t l3_lock;
@@ -1186,6 +1213,9 @@ struct rmon_ipv4_dst_offload {
 #define rswitch_undo_for_prev_ports(_priv, _rdev) \
 	list_for_each_entry_continue_reverse(_rdev, &_priv->port_list, list)
 
+#define rswitch_for_all_vports(_priv, _rvpb) \
+	list_for_each_entry(_rvpb, &_priv->vport_list, list)
+
 static void rswitch_update_l2_offload(struct rswitch_private *priv);
 static void rswitch_update_l3_offload(struct rswitch_private *priv);
 
@@ -1202,6 +1232,9 @@ static bool parallel_mode;
 module_param(parallel_mode, bool, 0644);
 MODULE_PARM_DESC(parallel_mode, "Operate simultaneously with Realtime core");
 
+/* FIXME - support multiple instances */
+static struct rswitch_private *the_rswitch_priv;
+
 #define RSWITCH_TIMEOUT_MS	1000
 
 #define rswitch_poll_timeout(addr, mask, expected) ({ \
@@ -3012,6 +3045,111 @@ static void rswitch_coma_init(struct rswitch_private *priv)
 	rswitch_write(priv, CABPPFLC0, CABPPFLC_INIT_VALUE);
 }
 
+
+static int rvp_be_open(struct net_device *ndev)
+{
+	struct rswitch_vport_backend *rvpb = netdev_priv(ndev);
+	struct rswitch_private *priv = rvpb->priv;
+	unsigned long flags;
+
+	spin_lock_irqsave(&priv->lock, flags);
+	rswitch_enadis_data_irq(priv, rvpb->tx_chain_index, true);
+	rswitch_enadis_data_irq(priv, rvpb->rx_chain_index, true);
+	spin_unlock_irqrestore(&priv->lock, flags);
+
+	napi_enable(&rvpb->napi);
+	netif_start_queue(ndev);
+
+	return 0;
+}
+
+static int rvp_be_close(struct net_device *ndev)
+{
+	struct rswitch_vport_backend *rvpb = netdev_priv(ndev);
+	struct rswitch_private *priv = rvpb->priv;
+	unsigned long flags;
+
+	napi_disable(&rvpb->napi);
+	netif_stop_queue(ndev);
+
+	spin_lock_irqsave(&priv->lock, flags);
+	rswitch_enadis_data_irq(priv, rvpb->tx_chain_index, false);
+	rswitch_enadis_data_irq(priv, rvpb->rx_chain_index, false);
+	spin_unlock_irqrestore(&priv->lock, flags);
+
+	return 0;
+}
+
+static int rvp_be_start_xmit(struct sk_buff *skb, struct net_device *ndev)
+{
+	struct rswitch_vport_backend *rvpb = netdev_priv(ndev);
+	struct rswitch_private *priv = rvpb->priv;
+	u64 info1_to_fe = INFO1_FMT | INFO1_IPV(GWCA_IPV_NUM) |
+			  INFO1_DV(BIT(GWCA_PORT)) |
+			  INFO1_CSD(GWCA_INDEX, rvpb->fe_rx_chain_index);
+	int ret, ring_ret;
+
+	ring_ret = rswitch_tx_ring_add_skb(rvpb->tx_ring, skb, 0, info1_to_fe);
+	if (ring_ret == 0) {
+		rswitch_kick_tx_queue(priv, rvpb->tx_chain_index);
+		ret = NETDEV_TX_OK;
+	} else {
+		if (ring_ret == -ENOSPC)
+			netif_stop_subqueue(ndev, 0);
+		ret = NETDEV_TX_BUSY;
+	}
+
+	return ret;
+}
+
+static int rvp_be_napi_poll(struct napi_struct *napi, int budget)
+{
+	struct net_device *ndev = napi->dev;
+	struct rswitch_vport_backend *rvpb = netdev_priv(ndev);
+	struct rswitch_private *priv = rvpb->priv;
+	struct sk_buff *skb;
+	int work_done = 0;
+	unsigned long flags;
+
+	if (rswitch_tx_ring_handle_complete(rvpb->tx_ring, &ndev->stats))
+		netif_wake_subqueue(ndev, 0);
+
+	while (work_done < budget) {
+		skb = rswitch_rx_ring_process(rvpb->rx_ring, false);
+		if (skb) {
+			skb->protocol = eth_type_trans(skb, ndev);
+			ndev->stats.rx_packets++;
+			ndev->stats.rx_bytes += skb->len;
+			napi_gro_receive(napi, skb);
+		}
+
+		work_done++;
+		if (!rswitch_rx_ring_need_process(rvpb->rx_ring))
+			break;
+	}
+
+	if (work_done < budget && napi_complete_done(napi, work_done)) {
+		spin_lock_irqsave(&priv->lock, flags);
+		rswitch_enadis_data_irq(priv, rvpb->tx_chain_index, true);
+		rswitch_enadis_data_irq(priv, rvpb->rx_chain_index, true);
+		spin_unlock_irqrestore(&priv->lock, flags);
+	}
+
+	return work_done;
+}
+
+static struct net_device_stats *rvp_be_get_stats(struct net_device *ndev)
+{
+	return &ndev->stats;
+}
+
+static const struct net_device_ops rvp_be_netdev_ops = {
+	.ndo_open = rvp_be_open,
+	.ndo_stop = rvp_be_close,
+	.ndo_start_xmit = rvp_be_start_xmit,
+	.ndo_get_stats = rvp_be_get_stats,
+};
+
 static bool rswitch_check_ack_irq(struct rswitch_private *priv,
 		u32 *status, int chain)
 {
@@ -3036,6 +3174,19 @@ static void rdev_interrupt(struct rswitch_device *rdev)
 	}
 }
 
+static void rvp_be_interrupt(struct rswitch_vport_backend *rvpb)
+{
+	struct rswitch_private *priv = rvpb->priv;
+
+	if (napi_schedule_prep(&rvpb->napi)) {
+		spin_lock(&priv->lock);
+		rswitch_enadis_data_irq(priv, rvpb->tx_chain_index, false);
+		rswitch_enadis_data_irq(priv, rvpb->rx_chain_index, false);
+		spin_unlock(&priv->lock);
+		__napi_schedule(&rvpb->napi);
+	}
+}
+
 static irqreturn_t rswitch_irq(int irq, void *dev_id)
 {
 	struct rswitch_private *priv = dev_id;
@@ -3043,6 +3194,7 @@ static irqreturn_t rswitch_irq(int irq, void *dev_id)
 	u32 irq_status[RSWITCH_NUM_IRQ_REGS];
 	int i;
 	struct rswitch_device *rdev;
+	struct rswitch_vport_backend *rvpb;
 	bool tx_irq, rx_irq;
 
 	for (i = 0; i < RSWITCH_NUM_IRQ_REGS; i++)
@@ -3060,6 +3212,17 @@ static irqreturn_t rswitch_irq(int irq, void *dev_id)
 		}
 	}
 
+	rswitch_for_all_vports(priv, rvpb) {
+		tx_irq = rswitch_check_ack_irq(priv, irq_status,
+				rvpb->tx_chain_index);
+		rx_irq = rswitch_check_ack_irq(priv, irq_status,
+				rvpb->rx_chain_index);
+		if (tx_irq || rx_irq) {
+			rvp_be_interrupt(rvpb);
+			ret = IRQ_HANDLED;
+		}
+	}
+
 	return ret;
 }
 
@@ -4561,6 +4724,248 @@ static void rdev_destroy(struct rswitch_device *rdev)
 	free_netdev(rdev->ndev);
 }
 
+int rswitch_vport_register(struct rswitch_vport_desc *rvpd)
+{
+	struct rswitch_private *priv = the_rswitch_priv;
+	unsigned int id;
+	struct net_device *ndev;
+	struct rswitch_vport_backend *rvpb;
+	int ret, ret2;
+
+	/* FIXME: locking */
+
+	if (parallel_mode)
+		return -ENOTSUPP;
+	if (!priv)
+		return -ENODEV;
+
+	if (rvpd->irq_index < 1 || rvpd->irq_index > 7)
+		return -EINVAL;
+
+	if (rvpd->id == RSWITCH_VPORT_ID_DYNAMIC) {
+		if (priv->vport_ids == ~0UL)
+			return -EBUSY;
+		id = ffz(priv->vport_ids);
+		rvpd->id = id;
+	} else {
+		if (rvpd->id >= 8 * sizeof(priv->vport_ids))
+			return -EINVAL;
+		if (priv->vport_ids & BIT(rvpd->id))
+			return -EBUSY;
+		id = rvpd->id;
+	}
+	priv->vport_ids |= BIT(rvpd->id);
+
+	ret = -ENOMEM;
+
+	ndev = alloc_etherdev(sizeof(*rvpb));
+	if (!ndev)
+		goto no_ndev;
+
+	SET_NETDEV_DEV(ndev, &priv->pdev->dev);
+	snprintf(ndev->name, IFNAMSIZ, "rvpback%d", id);
+	ndev->features = NETIF_F_NETNS_LOCAL;
+	ndev->netdev_ops = &rvp_be_netdev_ops;
+
+	ndev->addr_assign_type = NET_ADDR_STOLEN;
+	ether_addr_copy(ndev->dev_addr, priv->gwca.mac_addr);
+
+	rvpb = netdev_priv(ndev);
+	rvpb->rvpd = rvpd;
+	rvpb->id = id;
+	rvpb->priv = priv;
+	rvpb->ndev = ndev;
+
+	rvpd->rvpb = rvpb;
+
+	ret = -ENOSPC;
+
+	rvpb->tx_chain_index = rswitch_gwca_get(priv);
+	if (rvpb->tx_chain_index < 0)
+		goto no_be_tx_chain;
+	rvpb->rx_chain_index = rswitch_gwca_get(priv);
+	if (rvpb->rx_chain_index < 0)
+		goto no_be_rx_chain;
+	rvpb->fe_tx_chain_index = rswitch_gwca_get(priv);
+	if (rvpb->fe_tx_chain_index < 0)
+		goto no_fe_tx_chain;
+	rvpb->fe_rx_chain_index = rswitch_gwca_get(priv);
+	if (rvpb->fe_rx_chain_index < 0)
+		goto no_fe_rx_chain;
+
+	ret = -ENOMEM;
+
+	rvpb->tx_ring = rswitch_tx_ring_create(&priv->pdev->dev, 256);
+	if (!rvpb->tx_ring)
+		goto no_be_tx_ring;
+
+	rvpb->rx_ring = rswitch_rx_ring_create(&priv->pdev->dev, 256);
+	if (!rvpb->tx_ring)
+		goto no_be_rx_ring;
+
+	rswitch_gwca_attach_ring(priv, rvpb->tx_chain_index,
+			0, rswitch_tx_ring_dma_addr(rvpb->tx_ring),
+			GWDCC_DQT |	/* transmit queue */
+			GWDCC_DCP(GWCA_IPV_NUM) | /* transmit priority */
+			GWDCC_EDE);	/* extended descriptor */
+
+	rswitch_gwca_attach_ring(priv, rvpb->rx_chain_index,
+			0, rswitch_rx_ring_dma_addr(rvpb->rx_ring),
+			GWDCC_EDE |	/* extended descriptor */
+			GWDCC_ETS);	/* ... with timestamps */
+
+	/* FE rings are not yet attached, and observed iommu errors (attempt
+	 * to resolve address 0) when BE->FE send happened before FE is opened
+	 * and rings attached. Looks like need explicit chain initialization
+	 * so it notices EOS in BAT */
+	rswitch_gwca_modify(priv, GWDCC(rvpb->fe_rx_chain_index), 0, GWDCC_BALR);
+
+	if (!rvpd->fe_reg_base)
+		rvpd->fe_reg_base = priv->addr;
+	if (rvpd->fe_irq < 0) {
+		ret = rswitch_get_irq(priv, rvpd->irq_index);
+		if (ret < 0)
+			goto no_fe_irq;
+		rvpd->fe_irq = ret;
+	}
+	if (!rvpd->dma_dev)
+		rvpd->dma_dev = &priv->pdev->dev;
+
+	rvpd->info1_to_be = INFO1_FMT | INFO1_IPV(GWCA_IPV_NUM) |
+			    INFO1_DV(BIT(GWCA_PORT)) |
+			    INFO1_CSD(GWCA_INDEX, rvpb->rx_chain_index);
+	rvpd->info1_to_fwd = INFO1_IPV(GWCA_IPV_NUM);
+
+	rvpd->kick_tx_offset = GWTRC(rvpb->fe_tx_chain_index / 32);
+	rvpd->kick_tx_val = BIT(rvpb->fe_tx_chain_index % 32);
+
+	rvpd->tx_irq_ena_offset = GWDIE(rvpb->fe_tx_chain_index / 32);
+	rvpd->rx_irq_ena_offset = GWDIE(rvpb->fe_rx_chain_index / 32);
+	rvpd->tx_irq_dis_offset = GWDID(rvpb->fe_tx_chain_index / 32);
+	rvpd->rx_irq_dis_offset = GWDID(rvpb->fe_rx_chain_index / 32);
+	rvpd->tx_irq_status_offset = GWDIS(rvpb->fe_tx_chain_index / 32);
+	rvpd->rx_irq_status_offset = GWDIS(rvpb->fe_rx_chain_index / 32);
+	rvpd->tx_irq_bit = BIT(rvpb->fe_tx_chain_index % 32);
+	rvpd->rx_irq_bit = BIT(rvpb->fe_rx_chain_index % 32);
+
+	netif_napi_add(ndev, &rvpb->napi, rvp_be_napi_poll, 64);
+
+	ret = register_netdev(ndev);
+	if (ret)
+		goto no_ndev_register;
+
+	list_add_tail(&rvpb->list, &priv->vport_list);
+	return 0;
+
+no_ndev_register:
+	netif_napi_del(&rvpb->napi);
+
+no_fe_irq:
+	ret2 = rswitch_gwca_detach_ring(priv, rvpb->rx_chain_index);
+	if (ret2)
+		dev_err(&priv->pdev->dev,
+			"vport %u: be rx chain detach failed\n", id);
+
+	ret2 = rswitch_gwca_detach_ring(priv, rvpb->tx_chain_index);
+	if (ret2)
+		dev_err(&priv->pdev->dev,
+			"vport %u: be tx chain detach failed\n", id);
+
+	rswitch_rx_ring_destroy(rvpb->rx_ring);
+no_be_rx_ring:
+	rswitch_tx_ring_destroy(rvpb->tx_ring);
+no_be_tx_ring:
+	rswitch_gwca_put(priv, rvpb->fe_rx_chain_index);
+no_fe_rx_chain:
+	rswitch_gwca_put(priv, rvpb->fe_tx_chain_index);
+no_fe_tx_chain:
+	rswitch_gwca_put(priv, rvpb->rx_chain_index);
+no_be_rx_chain:
+	rswitch_gwca_put(priv, rvpb->tx_chain_index);
+no_be_tx_chain:
+	free_netdev(ndev);
+no_ndev:
+	priv->vport_ids &= ~BIT(id);
+	return ret;
+}
+EXPORT_SYMBOL(rswitch_vport_register);
+
+void rswitch_vport_unregister(struct rswitch_vport_desc *rvpd)
+{
+	struct rswitch_vport_backend *rvpb = rvpd->rvpb;
+	struct rswitch_private *priv = rvpb->priv;
+	unsigned int id = rvpb->id;
+	int ret;
+
+	list_del(&rvpb->list);
+
+	unregister_netdev(rvpb->ndev);
+	netif_napi_del(&rvpb->napi);
+
+	ret = rswitch_gwca_detach_ring(priv, rvpb->rx_chain_index);
+	if (ret)
+		dev_err(&priv->pdev->dev,
+			"vport %u: be rx chain detach failed\n", id);
+
+	ret = rswitch_gwca_detach_ring(priv, rvpb->tx_chain_index);
+	if (ret)
+		dev_err(&priv->pdev->dev,
+			"vport %u: be tx chain detach failed\n", id);
+
+	rswitch_rx_ring_destroy(rvpb->rx_ring);
+	rswitch_tx_ring_destroy(rvpb->tx_ring);
+
+	rswitch_gwca_put(priv, rvpb->fe_rx_chain_index);
+	rswitch_gwca_put(priv, rvpb->fe_tx_chain_index);
+	rswitch_gwca_put(priv, rvpb->rx_chain_index);
+	rswitch_gwca_put(priv, rvpb->tx_chain_index);
+
+	free_netdev(rvpb->ndev);
+
+	priv->vport_ids &= ~BIT(id);
+}
+EXPORT_SYMBOL(rswitch_vport_unregister);
+
+int rswitch_vport_attach_frontend_rings(struct rswitch_vport_desc *rvpd)
+{
+	struct rswitch_vport_backend *rvpb = rvpd->rvpb;
+	struct rswitch_private *priv = rvpb->priv;
+
+	rswitch_gwca_attach_ring(priv, rvpd->rvpb->fe_tx_chain_index,
+			rvpd->irq_index, rvpd->fe_tx_ring_dma,
+			GWDCC_DQT |	/* transmit queue */
+			GWDCC_DCP(GWCA_IPV_NUM) | /* transmit priority */
+			GWDCC_EDE);	/* extended descriptor */
+
+	rswitch_gwca_attach_ring(priv, rvpd->rvpb->fe_rx_chain_index,
+			rvpd->irq_index, rvpd->fe_rx_ring_dma,
+			GWDCC_EDE |	/* extended descriptor */
+			GWDCC_ETS);	/* ... with timestamps */
+
+	return 0;
+}
+EXPORT_SYMBOL(rswitch_vport_attach_frontend_rings);
+
+int rswitch_vport_detach_frontend_rings(struct rswitch_vport_desc *rvpd)
+{
+	struct rswitch_vport_backend *rvpb = rvpd->rvpb;
+	struct rswitch_private *priv = rvpb->priv;
+	int ret_rx, ret_tx;
+
+	ret_rx = rswitch_gwca_detach_ring(priv, rvpd->rvpb->fe_rx_chain_index);
+	if (ret_rx)
+		dev_err(&priv->pdev->dev,
+			"vport %u: fe rx chain detach failed\n", rvpd->id);
+
+	ret_tx = rswitch_gwca_detach_ring(priv, rvpd->rvpb->fe_tx_chain_index);
+	if (ret_tx)
+		dev_err(&priv->pdev->dev,
+			"vport %u: fe tx chain detach failed\n", rvpd->id);
+
+	return ret_rx ?: ret_tx;
+}
+EXPORT_SYMBOL(rswitch_vport_detach_frontend_rings);
+
 static int rswitch_probe(struct platform_device *pdev)
 {
 	struct rswitch_private *priv;
@@ -4644,6 +5049,7 @@ static int rswitch_probe(struct platform_device *pdev)
 	INIT_LIST_HEAD(&priv->port_list);
 	ports_dn = of_get_child_by_name(priv->pdev->dev.of_node, "ports");
 	if (!ports_dn) {
+		/* FIXME: handle virtual ports only? */
 		dev_err(&pdev->dev, "no ports subnode\n");
 		return -EINVAL;
 	}
@@ -4655,6 +5061,8 @@ static int rswitch_probe(struct platform_device *pdev)
 	if (list_empty(&priv->port_list))
 		dev_warn(&pdev->dev, "could not initialize any ports\n");
 
+	INIT_LIST_HEAD(&priv->vport_list);
+
 	if (!parallel_mode) {
 		pm_runtime_enable(&pdev->dev);
 		pm_runtime_get_sync(&pdev->dev);
@@ -4676,6 +5084,10 @@ static int rswitch_probe(struct platform_device *pdev)
 
 	device_set_wakeup_capable(&pdev->dev, 1);
 
+	if (!WARN_ON(the_rswitch_priv != NULL)) {
+		the_rswitch_priv = priv;
+	}
+
 	return 0;
 
 rswitch_register_notifiers_failed:
@@ -4703,6 +5115,10 @@ static int rswitch_remove(struct platform_device *pdev)
 	struct rswitch_private *priv = platform_get_drvdata(pdev);
 	struct rswitch_device *rdev;
 
+	if (!WARN_ON(the_rswitch_priv != priv)) {
+		the_rswitch_priv = NULL;
+	}
+
 	if (!parallel_mode)
 		rswitch_unregister_notifiers();
 
@@ -4720,6 +5136,9 @@ static int rswitch_remove(struct platform_device *pdev)
 		rdev_destroy(rdev);
 	}
 
+	/* FIXME: handle case if detaching while vports are remaining */
+	BUG_ON(!list_empty(&priv->vport_list));
+
 	return 0;
 }
 
@@ -4728,6 +5147,10 @@ static int __maybe_unused rswitch_suspend(struct device *dev)
 	struct rswitch_private *priv = dev_get_drvdata(dev);
 	struct rswitch_device *rdev;
 
+	/* FIXME */
+	if (!list_empty(&priv->vport_list))
+		return -EINVAL;
+
 	rswitch_for_all_ports(priv, rdev) {
 		struct net_device *ndev = rdev->ndev;
 
diff --git a/drivers/net/ethernet/renesas/rswitch_desc.h b/drivers/net/ethernet/renesas/rswitch_desc.h
index cd4fec1f6561..cf17fb43b28a 100644
--- a/drivers/net/ethernet/renesas/rswitch_desc.h
+++ b/drivers/net/ethernet/renesas/rswitch_desc.h
@@ -90,6 +90,8 @@ enum {
 #define INFO1_TSUN(val)		((u64)(val) << 8ULL)
 #define INFO1_IPV(prio)		((u64)(prio) << 28ULL)
 #define INFO1_DV(port_vector)	((u64)(port_vector) << 48ULL)
+#define INFO1_CSD(gwca_index, chain_index) \
+		((u64)(chain_index) << (32ULL + 8ULL * (gwca_index)))
 
 /* For timestamp descriptor in dptrl (Byte 4 to 7) */
 #define TS_DESC_TSUN(dptrl)	((dptrl) & GENMASK(7, 0))
diff --git a/drivers/net/ethernet/renesas/rswitch_lib.c b/drivers/net/ethernet/renesas/rswitch_lib.c
index 32004ed9243d..3b0d342308be 100644
--- a/drivers/net/ethernet/renesas/rswitch_lib.c
+++ b/drivers/net/ethernet/renesas/rswitch_lib.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 
+#include <linux/etherdevice.h>
 #include <linux/module.h>
-#include <linux/netdevice.h>
 
 #include "rswitch_desc.h"
 #include "rswitch_lib.h"
@@ -426,4 +426,264 @@ void rswitch_rx_ring_destroy(struct rswitch_rx_ring *ring)
 }
 EXPORT_SYMBOL(rswitch_rx_ring_destroy);
 
+
+struct rswitch_vport_frontend {
+	struct rswitch_vport_desc *rvpd;
+
+	struct net_device *ndev;
+	struct napi_struct napi;
+
+	struct rswitch_tx_ring *tx_ring;
+	struct rswitch_rx_ring *rx_ring;
+
+	spinlock_t lock;
+};
+
+#define RVP_READ(rvpd, field) \
+	readl(rvpd->fe_reg_base + rvpd->field ## _offset)
+#define RVP_WRITE(rvpd, field, val) \
+	writel((val), rvpd->fe_reg_base + rvpd->field ## _offset)
+
+static irqreturn_t rvpf_irq_handler(int irq, void *data)
+{
+	struct rswitch_vport_frontend *rvpf = data;
+	struct rswitch_vport_desc *rvpd = rvpf->rvpd;
+	u32 tx_status, rx_status;
+
+	tx_status = RVP_READ(rvpd, tx_irq_status) & rvpd->tx_irq_bit;
+	if (tx_status)
+		RVP_WRITE(rvpd, tx_irq_status, tx_status);
+
+	rx_status = RVP_READ(rvpd, rx_irq_status) & rvpd->rx_irq_bit;
+	if (rx_status)
+		RVP_WRITE(rvpd, rx_irq_status, rx_status);
+
+	if (!tx_status && !rx_status)
+		return IRQ_NONE;
+
+	if (napi_schedule_prep(&rvpf->napi)) {
+		spin_lock(&rvpf->lock);
+		RVP_WRITE(rvpd, tx_irq_dis, rvpd->tx_irq_bit);
+		RVP_WRITE(rvpd, rx_irq_dis, rvpd->rx_irq_bit);
+		spin_unlock(&rvpf->lock);
+
+		__napi_schedule(&rvpf->napi);
+	}
+
+	return IRQ_HANDLED;
+}
+
+static int rvpf_open(struct net_device *ndev)
+{
+	struct rswitch_vport_frontend *rvpf = netdev_priv(ndev);
+	struct rswitch_vport_desc *rvpd = rvpf->rvpd;
+	int ret;
+	unsigned long flags;
+	bool can_destroy_rings = true;
+
+	rvpf->tx_ring = rswitch_tx_ring_create(rvpd->dma_dev, 256);
+	if (!rvpf->tx_ring) {
+		netdev_err(ndev, "could not create tx ring\n");
+		ret = -ENOMEM;
+		goto tx_ring_create_failed;
+	}
+	rvpd->fe_tx_ring_dma = rswitch_tx_ring_dma_addr(rvpf->tx_ring);
+
+	rvpf->rx_ring = rswitch_rx_ring_create(rvpd->dma_dev, 256);
+	if (!rvpf->rx_ring) {
+		netdev_err(ndev, "could not create rx ring\n");
+		ret = -ENOMEM;
+		goto rx_ring_create_failed;
+	}
+	rvpd->fe_rx_ring_dma = rswitch_rx_ring_dma_addr(rvpf->rx_ring);
+
+	ret = rvpd->ops->attach_frontend_rings(rvpd);
+	if (ret) {
+		netdev_err(ndev, "could not attach rings\n");
+		goto attach_rings_failed;
+	}
+
+	ret = request_irq(rvpd->fe_irq, rvpf_irq_handler,
+			IRQF_SHARED, netdev_name(ndev), rvpf);
+	if (ret) {
+		netdev_err(ndev, "could not request irq");
+		goto request_irq_failed;
+	}
+
+	spin_lock_irqsave(&rvpf->lock, flags);
+	RVP_WRITE(rvpd, tx_irq_ena, rvpd->tx_irq_bit);
+	RVP_WRITE(rvpd, rx_irq_ena, rvpd->rx_irq_bit);
+	spin_unlock_irqrestore(&rvpf->lock, flags);
+
+	netif_start_queue(ndev);
+	napi_enable(&rvpf->napi);
+
+	return 0;
+
+request_irq_failed:
+	can_destroy_rings = (rvpd->ops->detach_frontend_rings(rvpd) == 0);
+attach_rings_failed:
+	if (can_destroy_rings)
+		rswitch_rx_ring_destroy(rvpf->rx_ring);
+rx_ring_create_failed:
+	if (can_destroy_rings)
+		rswitch_tx_ring_destroy(rvpf->tx_ring);
+tx_ring_create_failed:
+	return ret;
+}
+
+static int rvpf_close(struct net_device *ndev)
+{
+	struct rswitch_vport_frontend *rvpf = netdev_priv(ndev);
+	struct rswitch_vport_desc *rvpd = rvpf->rvpd;
+	unsigned long flags;
+	bool can_destroy_rings;
+
+	napi_disable(&rvpf->napi);
+	netif_stop_queue(ndev);
+
+	spin_lock_irqsave(&rvpf->lock, flags);
+	RVP_WRITE(rvpd, tx_irq_dis, rvpd->tx_irq_bit);
+	RVP_WRITE(rvpd, rx_irq_dis, rvpd->rx_irq_bit);
+	spin_unlock_irqrestore(&rvpf->lock, flags);
+
+	free_irq(rvpd->fe_irq, rvpf);
+
+	can_destroy_rings = (rvpd->ops->detach_frontend_rings(rvpd) == 0);
+	if (can_destroy_rings) {
+		rswitch_rx_ring_destroy(rvpf->rx_ring);
+		rswitch_tx_ring_destroy(rvpf->tx_ring);
+	} else {
+		netdev_warn(ndev, "rings not detached - skipping destroy\n");
+	}
+
+	return 0;
+}
+
+static int rvpf_start_xmit(struct sk_buff *skb, struct net_device *ndev)
+{
+	struct rswitch_vport_frontend *rvpf = netdev_priv(ndev);
+	struct rswitch_vport_desc *rvpd = rvpf->rvpd;
+	int ret, ring_ret;
+
+	ring_ret = rswitch_tx_ring_add_skb(rvpf->tx_ring, skb, 0, rvpd->info1_to_be);
+	if (ring_ret == 0) {
+		RVP_WRITE(rvpd, kick_tx, rvpd->kick_tx_val);
+		ret = NETDEV_TX_OK;
+	} else {
+		if (ring_ret == -ENOSPC)
+			netif_stop_subqueue(ndev, 0);
+		ret = NETDEV_TX_BUSY;
+	}
+
+	return ret;
+}
+
+static int rvpf_napi_poll(struct napi_struct *napi, int budget)
+{
+	struct net_device *ndev = napi->dev;
+	struct rswitch_vport_frontend *rvpf = netdev_priv(ndev);
+	struct rswitch_vport_desc *rvpd = rvpf->rvpd;
+	struct sk_buff *skb;
+	int work_done = 0;
+	unsigned long flags;
+
+	if (rswitch_tx_ring_handle_complete(rvpf->tx_ring, &ndev->stats))
+		netif_wake_subqueue(ndev, 0);
+
+	while (work_done < budget) {
+		skb = rswitch_rx_ring_process(rvpf->rx_ring, false);
+		if (skb) {
+			skb->protocol = eth_type_trans(skb, ndev);
+			ndev->stats.rx_packets++;
+			ndev->stats.rx_bytes += skb->len;
+			napi_gro_receive(napi, skb);
+		}
+
+		work_done++;
+		if (!rswitch_rx_ring_need_process(rvpf->rx_ring))
+			break;
+	}
+
+	if (work_done < budget && napi_complete_done(napi, work_done)) {
+		spin_lock_irqsave(&rvpf->lock, flags);
+		RVP_WRITE(rvpd, tx_irq_ena, rvpd->tx_irq_bit);
+		RVP_WRITE(rvpd, rx_irq_ena, rvpd->rx_irq_bit);
+		spin_unlock_irqrestore(&rvpf->lock, flags);
+	}
+
+	return work_done;
+}
+
+static struct net_device_stats *rvpf_get_stats(struct net_device *ndev)
+{
+	return &ndev->stats;
+}
+
+static const struct net_device_ops rvpf_netdev_ops = {
+	.ndo_open = rvpf_open,
+	.ndo_stop = rvpf_close,
+	.ndo_start_xmit = rvpf_start_xmit,
+	.ndo_get_stats = rvpf_get_stats,
+};
+
+int rswitch_vport_frontend_run(struct rswitch_vport_desc *rvpd)
+{
+	struct net_device *ndev;
+	struct rswitch_vport_frontend *rvpf;
+	int ret;
+
+	ndev = alloc_etherdev(sizeof(*rvpf));
+	if (!ndev) {
+		printk(KERN_ERR "could not alloc frontend netdev\n");
+		return -ENOMEM;
+	}
+
+	SET_NETDEV_DEV(ndev, rvpd->dma_dev);
+	ndev->netdev_ops = &rvpf_netdev_ops;
+	eth_hw_addr_random(ndev);
+
+	rvpf = netdev_priv(ndev);
+	rvpf->ndev = ndev;
+	rvpf->rvpd = rvpd;
+	spin_lock_init(&rvpf->lock);
+
+	ret = rvpd->ops->vport_register(rvpd);
+	if (ret) {
+		printk(KERN_ERR "could not register rswitch vport\n");
+		goto rvp_register_failed;
+	}
+
+	snprintf(ndev->name, IFNAMSIZ, "rvpfront%d", rvpd->id);
+	netif_napi_add(ndev, &rvpf->napi, rvpf_napi_poll, 64);
+
+	ret = register_netdev(ndev);
+	if (ret) {
+		netdev_err(ndev, "could not register\n");
+		goto netdev_register_failed;
+	}
+
+	rvpd->rvpf = rvpf;
+	return 0;
+
+netdev_register_failed:
+	netif_napi_del(&rvpf->napi);
+	rvpd->ops->vport_unregister(rvpd);
+rvp_register_failed:
+	free_netdev(ndev);
+	return ret;
+}
+EXPORT_SYMBOL(rswitch_vport_frontend_run);
+
+void rswitch_vport_frontend_stop(struct rswitch_vport_desc *rvpd)
+{
+	struct rswitch_vport_frontend *rvpf = rvpd->rvpf;
+
+	unregister_netdev(rvpf->ndev);
+	netif_napi_del(&rvpf->napi);
+	rvpd->ops->vport_unregister(rvpd);
+	free_netdev(rvpf->ndev);
+}
+EXPORT_SYMBOL(rswitch_vport_frontend_stop);
+
 MODULE_LICENSE("GPL v2");
diff --git a/drivers/net/ethernet/renesas/rswitch_lib.h b/drivers/net/ethernet/renesas/rswitch_lib.h
index 1b9a3c1b996f..e0c5c35593bf 100644
--- a/drivers/net/ethernet/renesas/rswitch_lib.h
+++ b/drivers/net/ethernet/renesas/rswitch_lib.h
@@ -3,7 +3,8 @@
 #ifndef RSWITCH_LIB_H
 #define RSWITCH_LIB_H
 
-#include <linux/types.h>
+#include <linux/netdevice.h>
+
 
 struct rswitch_tx_ring;
 struct rswitch_rx_ring;
@@ -90,4 +91,84 @@ struct sk_buff *rswitch_rx_ring_process(struct rswitch_rx_ring *ring,
  */
 bool rswitch_rx_ring_need_process(struct rswitch_rx_ring *ring);
 
+
+struct rswitch_private;
+struct rswitch_vport_backend;
+struct rswitch_vport_frontend;
+
+struct rswitch_vport_desc;
+struct rswitch_vport_desc_ops {
+	int (*vport_register)(struct rswitch_vport_desc *rvpd);
+	void (*vport_unregister)(struct rswitch_vport_desc *rvpd);
+
+	int (*attach_frontend_rings)(struct rswitch_vport_desc *rvpd);
+	int (*detach_frontend_rings)(struct rswitch_vport_desc *rvpd);
+};
+
+struct rswitch_vport_desc {
+	const struct rswitch_vport_desc_ops *ops;
+
+	/* Requested vport id, or RSWITCH_VPORT_ID_DYNAMIC for allocation */
+	unsigned int id;
+#define RSWITCH_VPORT_ID_DYNAMIC  ~(0u)
+
+	/* GWCA interrupt index to deliver frontend interrupts */
+	unsigned int irq_index;
+
+	/* Rswitch register base, valid at frontend side
+	 * (can be set to NULL if frontend driver and rswitch driver are in the
+	 * same kernel) */
+	void __iomem *fe_reg_base;
+
+	/* Interrupt number to deliver frontend interrupts, requestable at
+	 * frontend side
+	 * (can be set to -1 if frontend driver and rswitch driver are in the
+	 * same kernel) */
+	int fe_irq;
+
+	/* Device struct for DMA operations on frontend chains
+	 * (can be set to NULL if frontend driver and rswitch driver are in the
+	 * same kernel) */
+	struct device *dma_dev;
+
+	/* Rswitch driver's private pointer */
+	struct rswitch_vport_backend *rvpb;
+
+	/* Rswitch lib's private pointer */
+	struct rswitch_vport_frontend *rvpf;
+
+	/* Frontend Tx and Rx DMA addresses, accessible via the above OSID
+	 * - filled by lib, used by rswitch driver */
+	dma_addr_t fe_tx_ring_dma;
+	dma_addr_t fe_rx_ring_dma;
+
+	/* Operation details - filled by rswitch driver, used by lib */
+
+	u64 info1_to_be;	/* info1 for sending to backend */
+	u64 info1_to_fwd;	/* info1 for sending to switch forwarder */
+
+	u32 kick_tx_offset;	/* register to kick tx ring */
+	u32 kick_tx_val;	/* value to kick tx ring */
+
+	u32 tx_irq_ena_offset, rx_irq_ena_offset;
+	u32 tx_irq_dis_offset, rx_irq_dis_offset;
+	u32 tx_irq_status_offset, rx_irq_status_offset;
+	u32 tx_irq_bit, rx_irq_bit;
+};
+
+/* Provide full lifecycle of vport frontend device */
+int rswitch_vport_frontend_run(struct rswitch_vport_desc *rvpd);
+void rswitch_vport_frontend_stop(struct rswitch_vport_desc *rvpd);
+
+
+/* Rswitch driver exports */
+
+/* Vport registration
+ * (FIXME: support multiple rswitch instances)*/
+int rswitch_vport_register(struct rswitch_vport_desc *rvpd);
+void rswitch_vport_unregister(struct rswitch_vport_desc *rvpd);
+
+int rswitch_vport_attach_frontend_rings(struct rswitch_vport_desc *rvpd);
+int rswitch_vport_detach_frontend_rings(struct rswitch_vport_desc *rvpd);
+
 #endif
diff --git a/drivers/net/ethernet/renesas/rswitch_vport_local.c b/drivers/net/ethernet/renesas/rswitch_vport_local.c
new file mode 100644
index 000000000000..98f39f61bb99
--- /dev/null
+++ b/drivers/net/ethernet/renesas/rswitch_vport_local.c
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/module.h>
+
+#include "rswitch_lib.h"
+
+#define VPORT_NR	2
+static struct rswitch_vport_desc *rvpd_array[VPORT_NR];
+
+#define RVPL_IRQ_INDEX 7
+
+static const struct rswitch_vport_desc_ops rvpl_ops = {
+	.vport_register = rswitch_vport_register,
+	.vport_unregister = rswitch_vport_unregister,
+	.attach_frontend_rings = rswitch_vport_attach_frontend_rings,
+	.detach_frontend_rings = rswitch_vport_detach_frontend_rings,
+};
+
+static struct rswitch_vport_desc *rvpl_create(void)
+{
+	struct rswitch_vport_desc *rvpd;
+	int ret;
+
+	rvpd = kzalloc(sizeof(*rvpd), GFP_KERNEL);
+	if (!rvpd)
+		return ERR_PTR(-ENOMEM);
+
+	rvpd->ops = &rvpl_ops;
+	rvpd->id = RSWITCH_VPORT_ID_DYNAMIC;
+	rvpd->irq_index = RVPL_IRQ_INDEX;
+	rvpd->fe_irq = -1;
+
+	ret = rswitch_vport_frontend_run(rvpd);
+	if (ret) {
+		kfree(rvpd);
+		return ERR_PTR(ret);
+	}
+
+	return rvpd;
+}
+
+static void rvpl_remove(struct rswitch_vport_desc *rvpd)
+{
+	rswitch_vport_frontend_stop(rvpd);
+	kfree(rvpd);
+}
+
+static int rvpl_init(void)
+{
+	unsigned int i;
+	int ret;
+
+	for (i = 0; i < VPORT_NR; i++) {
+		rvpd_array[i] = rvpl_create();
+		if (IS_ERR(rvpd_array[i])) {
+			ret = PTR_ERR(rvpd_array[i]);
+			goto failed;
+		}
+	}
+
+	return 0;
+
+	do {
+		i--;
+		rvpl_remove(rvpd_array[i]);
+failed:
+		;
+	} while (i > 0);
+
+	return ret;
+}
+
+static void rvpl_exit(void)
+{
+	unsigned int i;
+
+	for (i = 0; i < VPORT_NR; i++)
+		rvpl_remove(rvpd_array[i]);
+}
+
+module_init(rvpl_init);
+module_exit(rvpl_exit);
+MODULE_LICENSE("GPL v2");
-- 
2.25.1

