From 2000c0688d62c46c796df225ddca7f15bcbfa0dc Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 14 Oct 2024 10:10:13 +0500
Subject: [PATCH 155/157] net: ethernet: renesas: rswitch: xen frontend/backend
 drivers for vports

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/drivers/net/ethernet/renesas/Kconfig b/drivers/net/ethernet/renesas/Kconfig
index 9ba5e37f899c..253c395096f1 100644
--- a/drivers/net/ethernet/renesas/Kconfig
+++ b/drivers/net/ethernet/renesas/Kconfig
@@ -66,6 +66,16 @@ config RENESAS_ETHER_SWITCH_VPORT_LOCAL
 	tristate "Local rswitch virtual port"
 	depends on RENESAS_ETHER_SWITCH
 
+config RENESAS_ETHER_SWITCH_XENBACK
+	tristate "Xen backend device for rswitch virtual port"
+	depends on RENESAS_ETHER_SWITCH && XEN_BACKEND
+
+config RENESAS_ETHER_SWITCH_XENFRONT
+	tristate "Xen frontend device for rswitch virtual port"
+	depends on XEN
+	select XEN_XENBUS_FRONTEND
+	select RENESAS_ETHER_SWITCH_LIB
+
 config RTSN
         tristate "Renesas Ethernet-TSN support"
         depends on ARCH_RENESAS || COMPILE_TEST
diff --git a/drivers/net/ethernet/renesas/Makefile b/drivers/net/ethernet/renesas/Makefile
index f34d12211477..3715b24b1e0d 100644
--- a/drivers/net/ethernet/renesas/Makefile
+++ b/drivers/net/ethernet/renesas/Makefile
@@ -14,3 +14,5 @@ obj-$(CONFIG_RTSN) += rtsn.o
 obj-$(CONFIG_RENESAS_ETHER_SWITCH_LIB) += rswitch_lib.o
 obj-$(CONFIG_RENESAS_ETHER_SWITCH) += rswitch.o
 obj-$(CONFIG_RENESAS_ETHER_SWITCH_VPORT_LOCAL) += rswitch_vport_local.o
+obj-$(CONFIG_RENESAS_ETHER_SWITCH_XENBACK) += rswitch_vport_xenback.o
+obj-$(CONFIG_RENESAS_ETHER_SWITCH_XENFRONT) += rswitch_vport_xenfront.o
diff --git a/drivers/net/ethernet/renesas/rswitch.c b/drivers/net/ethernet/renesas/rswitch.c
index 1da52d47707b..f969e58938d3 100644
--- a/drivers/net/ethernet/renesas/rswitch.c
+++ b/drivers/net/ethernet/renesas/rswitch.c
@@ -810,6 +810,8 @@ enum rswitch_gwca_mode {
 #define GWARIRM_ARIOG	BIT(0)
 #define GWARIRM_ARR	BIT(1)
 
+#define GWDCC_OSID_MASK		GENMASK(30, 28)
+#define GWDCC_OSID(osid)	FIELD_PREP(GWDCC_OSID_MASK, (osid))
 #define GWDCC_BALR		BIT(24)
 #define GWDCC_DCP_MASK		GENMASK(18, 16)
 #define GWDCC_DCP(prio)		FIELD_PREP(GWDCC_DCP_MASK, (prio))
@@ -1263,7 +1265,6 @@ static void rmon_handle_l3_learning(struct rswitch_route_monitor *rmon,
 		struct sk_buff *skb);
 static void rmon_cleanup_ipv4_offloads_all(struct rswitch_route_monitor *rmon);
 
-
 static int num_etha_ports = 3;
 module_param(num_etha_ports, int, 0644);
 MODULE_PARM_DESC(num_etha_ports, "Number of using ETHA ports");
@@ -5163,6 +5164,8 @@ int rswitch_vport_register(struct rswitch_vport_desc *rvpd)
 
 	if (rvpd->irq_index < 1 || rvpd->irq_index > 7)
 		return -EINVAL;
+	if (rvpd->osid > 7)
+		return -EINVAL;
 
 	if (rvpd->id == RSWITCH_VPORT_ID_DYNAMIC) {
 		if (priv->vport_ids == ~0UL)
@@ -5298,8 +5301,8 @@ int rswitch_vport_register(struct rswitch_vport_desc *rvpd)
 	if (list_empty(&priv->vport_list))
 		rswitch_vport_cleanup_shared_rx_ring(priv);
 
-no_shared_rx_ring:
 no_fe_irq:
+no_shared_rx_ring:
 	ret2 = rswitch_gwca_detach_ring(priv, rvpb->rx_chain_index);
 	if (ret2)
 		dev_err(&priv->pdev->dev,
@@ -5379,12 +5382,14 @@ int rswitch_vport_attach_frontend_rings(struct rswitch_vport_desc *rvpd)
 
 	rswitch_gwca_attach_ring(priv, rvpd->rvpb->fe_tx_chain_index,
 			rvpd->irq_index, rvpd->fe_tx_ring_dma,
+			GWDCC_OSID(rvpd->osid) |
 			GWDCC_DQT |	/* transmit queue */
 			GWDCC_DCP(GWCA_IPV_NUM) | /* transmit priority */
 			GWDCC_EDE);	/* extended descriptor */
 
 	rswitch_gwca_attach_ring(priv, rvpd->rvpb->fe_rx_chain_index,
 			rvpd->irq_index, rvpd->fe_rx_ring_dma,
+			GWDCC_OSID(rvpd->osid) |
 			GWDCC_EDE |	/* extended descriptor */
 			GWDCC_ETS);	/* ... with timestamps */
 
diff --git a/drivers/net/ethernet/renesas/rswitch_lib.h b/drivers/net/ethernet/renesas/rswitch_lib.h
index c0962ae85da2..346c9e523cb2 100644
--- a/drivers/net/ethernet/renesas/rswitch_lib.h
+++ b/drivers/net/ethernet/renesas/rswitch_lib.h
@@ -116,6 +116,9 @@ struct rswitch_vport_desc {
 	/* GWCA interrupt index to deliver frontend interrupts */
 	unsigned int irq_index;
 
+	/* OSID for GWCA operation with frontend chains */
+	unsigned int osid;
+
 	/* Rswitch register base, valid at frontend side
 	 * (can be set to NULL if frontend driver and rswitch driver are in the
 	 * same kernel) */
@@ -138,6 +141,9 @@ struct rswitch_vport_desc {
 	/* Rswitch lib's private pointer */
 	struct rswitch_vport_frontend *rvpf;
 
+	/* Frontend driver's private pointer */
+	void *fe_drv_priv;
+
 	/* Frontend Tx and Rx DMA addresses, accessible via the above OSID
 	 * - filled by lib, used by rswitch driver */
 	dma_addr_t fe_tx_ring_dma;
diff --git a/drivers/net/ethernet/renesas/rswitch_vport_xen.h b/drivers/net/ethernet/renesas/rswitch_vport_xen.h
new file mode 100644
index 000000000000..b3e5a7033698
--- /dev/null
+++ b/drivers/net/ethernet/renesas/rswitch_vport_xen.h
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#ifndef RSWITCH_VPORT_XEN_H
+#define RSWITCH_VPORT_XEN_H
+
+#include <xen/interface/event_channel.h>
+#include "rswitch_lib.h"
+
+struct rswitch_vport_xen_shm {
+	struct rswitch_vport_desc rvpd;
+
+	evtchn_port_t be_to_fe_chn;
+	evtchn_port_t fe_to_be_chn;
+
+	/* Command protocol:
+	 * - sender writes command code,
+	 * - interrupt handler in receiver writes BE_COMMAND_RECEIVED
+	 *   (needed for multiple interrupt handler calls not caused multiple
+	 *   command execution)
+	 * - after completing the command, receiver writes BE_COMMAND_DONE
+	 */
+	int fe_to_be_command;
+#define BE_COMMAND_DONE		1
+#define BE_COMMAND_IN_PROGRESS	2
+#define BE_COMMAND_REGISTER	3
+#define BE_COMMAND_UNREGISTER	4
+#define BE_COMMAND_ATTACH_RINGS	5
+#define BE_COMMAND_DETACH_RINGS	6
+	int fe_to_be_result;
+};
+
+#endif
diff --git a/drivers/net/ethernet/renesas/rswitch_vport_xenback.c b/drivers/net/ethernet/renesas/rswitch_vport_xenback.c
new file mode 100644
index 000000000000..f362646eb4a6
--- /dev/null
+++ b/drivers/net/ethernet/renesas/rswitch_vport_xenback.c
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/module.h>
+#include <xen/events.h>
+#include <xen/grant_table.h>
+#include <xen/xenbus.h>
+
+#include "rswitch_vport_xen.h"
+
+struct rvp_xenback {
+	struct xenbus_device *xdev;
+
+	struct rswitch_vport_xen_shm *shm;
+	int shm_gref;
+	int from_fe_irq;
+
+	struct work_struct work;
+	int work_command;
+
+	bool rvp_registered;
+};
+
+static void rvp_xenback_comm_work(struct work_struct *work)
+{
+	struct rvp_xenback *rvpx = container_of(work, typeof(*rvpx), work);
+	struct rswitch_vport_desc *rvpd = &rvpx->shm->rvpd;
+	int ret = 0;
+
+	switch (rvpx->work_command) {
+	case BE_COMMAND_REGISTER:
+		ret = rswitch_vport_register(rvpd);
+		if (ret == 0)
+			rvpx->rvp_registered = true;
+		break;
+	case BE_COMMAND_UNREGISTER:
+		if (rvpx->rvp_registered) {
+			rswitch_vport_unregister(rvpd);
+			rvpx->rvp_registered = false;
+		}
+		break;
+	case BE_COMMAND_ATTACH_RINGS:
+		if (rvpx->rvp_registered)
+			ret = rswitch_vport_attach_frontend_rings(rvpd);
+		else
+			ret = -EIO;
+		break;
+	case BE_COMMAND_DETACH_RINGS:
+		if (rvpx->rvp_registered)
+			ret = rswitch_vport_detach_frontend_rings(rvpd);
+		else
+			ret = -EIO;
+		break;
+	default:
+		ret = -EOPNOTSUPP;
+		break;
+	}
+
+	rvpx->shm->fe_to_be_result = ret;
+	wmb();
+	rvpx->shm->fe_to_be_command = BE_COMMAND_DONE;
+	notify_remote_via_evtchn(rvpx->shm->be_to_fe_chn);
+}
+
+static irqreturn_t rvp_xenback_comm_interrupt(int irq, void *data)
+{
+	struct rvp_xenback *rvpx = data;
+
+	if (rvpx->shm->fe_to_be_command > BE_COMMAND_IN_PROGRESS) {
+		rvpx->work_command = rvpx->shm->fe_to_be_command;
+		rvpx->shm->fe_to_be_command = BE_COMMAND_IN_PROGRESS;
+		schedule_work(&rvpx->work);
+	}
+
+	xen_irq_lateeoi(irq, 0);
+	return IRQ_HANDLED;
+}
+
+static void rvp_xenback_frontend_changed(struct xenbus_device *dev,
+		enum xenbus_state frontend_state)
+{
+	struct rvp_xenback *rvpx = dev_get_drvdata(&dev->dev);
+	int ret;
+
+	if (frontend_state == XenbusStateConnected &&
+	    dev->state != XenbusStateConnected) {
+		ret = bind_interdomain_evtchn_to_irqhandler_lateeoi(
+				dev->otherend_id,
+				rvpx->shm->fe_to_be_chn,
+				rvp_xenback_comm_interrupt, 0,
+				"rvp_xen_comm", rvpx);
+		if (ret < 0) {
+			xenbus_dev_fatal(dev, ret, "bind comm handler failed");
+			xenbus_switch_state(dev, XenbusStateClosed);
+		} else {
+			rvpx->from_fe_irq = ret;
+			xenbus_switch_state(dev, XenbusStateConnected);
+		}
+	}
+
+	if (frontend_state != XenbusStateConnected &&
+	    dev->state == XenbusStateConnected) {
+		if (rvpx->rvp_registered) {
+			rswitch_vport_unregister(&rvpx->shm->rvpd);
+			rvpx->rvp_registered = false;
+		}
+
+		unbind_from_irqhandler(rvpx->from_fe_irq, rvpx);
+		/* FXIME: reinit? */
+		xenbus_switch_state(dev, XenbusStateClosed);
+	}
+}
+
+static int rvp_xenback_probe(struct xenbus_device *dev,
+		const struct xenbus_device_id *id)
+{
+	struct rvp_xenback *rvpx;
+	int ret;
+
+	rvpx = devm_kzalloc(&dev->dev, sizeof(*rvpx), GFP_KERNEL);
+	if (!rvpx) {
+		ret = -ENOMEM;
+		xenbus_dev_fatal(dev, ret, "allocation failed");
+		goto rvpx_alloc_failed;
+	}
+
+	rvpx->xdev = dev;
+	dev_set_drvdata(&dev->dev, rvpx);
+	INIT_WORK(&rvpx->work, rvp_xenback_comm_work);
+
+	rvpx->shm = (struct rswitch_vport_xen_shm *)
+		__get_free_page(GFP_KERNEL);
+	if (!rvpx->shm) {
+		ret = -ENOMEM;
+		xenbus_dev_fatal(dev, ret, "shm allocation failed");
+		goto shm_alloc_failed;
+	}
+	memset(rvpx->shm, 0, sizeof(*rvpx->shm));
+
+	ret = gnttab_grant_foreign_access(dev->otherend_id,
+			virt_to_gfn(rvpx->shm), 0);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "grant shm access failed");
+		goto shm_grant_failed;
+	}
+	rvpx->shm_gref = ret;
+	ret = xenbus_printf(XBT_NIL, dev->nodename, "shm-gref",
+			"%u", rvpx->shm_gref);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "shm gref publish failed");
+		goto shm_gref_publish_failed;
+	}
+
+	ret = xenbus_alloc_evtchn(dev, &rvpx->shm->be_to_fe_chn);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "be_to_fe channel alloc failed");
+		goto channel_alloc_failed;
+	}
+
+	xenbus_switch_state(dev, XenbusStateInitWait);
+	return 0;
+
+channel_alloc_failed:
+shm_gref_publish_failed:
+	gnttab_end_foreign_access(rvpx->shm_gref, 0, (unsigned long)rvpx->shm);
+shm_grant_failed:
+	put_page(virt_to_page((unsigned long)rvpx->shm));
+shm_alloc_failed:
+rvpx_alloc_failed:
+	xenbus_switch_state(dev, XenbusStateClosed);
+return ret;
+}
+
+static int rvp_xenback_remove(struct xenbus_device *dev)
+{
+	struct rvp_xenback *rvpx = dev_get_drvdata(&dev->dev);
+
+	xenbus_switch_state(dev, XenbusStateClosing);
+
+	xenbus_free_evtchn(dev, rvpx->shm->be_to_fe_chn);
+	gnttab_end_foreign_access(rvpx->shm_gref, 0, (unsigned long)rvpx->shm);
+	put_page(virt_to_page((unsigned long)rvpx->shm));
+
+	xenbus_switch_state(dev, XenbusStateClosed);
+	return 0;
+}
+
+static const struct xenbus_device_id rvp_xenback_ids[] = {
+	{ "rswitch_vport" },
+	{ "" }
+};
+
+static struct xenbus_driver rvp_xenback_driver = {
+	.ids = rvp_xenback_ids,
+	.probe = rvp_xenback_probe,
+	.remove = rvp_xenback_remove,
+	.otherend_changed = rvp_xenback_frontend_changed,
+};
+
+static int __init rvp_xenback_init(void)
+{
+	if (!xen_domain())
+		return -ENODEV;
+
+	return xenbus_register_backend(&rvp_xenback_driver);
+}
+module_init(rvp_xenback_init);
+
+static void __exit rvp_xenback_exit(void)
+{
+	xenbus_unregister_driver(&rvp_xenback_driver);
+}
+module_exit(rvp_xenback_exit);
+
+MODULE_LICENSE("GPL");
diff --git a/drivers/net/ethernet/renesas/rswitch_vport_xenfront.c b/drivers/net/ethernet/renesas/rswitch_vport_xenfront.c
new file mode 100644
index 000000000000..8d4dc187f848
--- /dev/null
+++ b/drivers/net/ethernet/renesas/rswitch_vport_xenfront.c
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/of_device.h>
+#include <xen/events.h>
+#include <xen/grant_table.h>
+#include <xen/xenbus.h>
+
+#include "rswitch_vport_xen.h"
+
+struct rvp_xenfront {
+	struct xenbus_device *xdev;
+
+	struct page *shm_page;
+	struct rswitch_vport_xen_shm *shm;
+	int shm_gref;
+	int from_be_irq;
+
+	struct wait_queue_head wait;
+
+	bool rvpf_running;
+};
+
+static struct device *rswitch_dev;
+static void __iomem *rswitch_regs;
+static int rswitch_irq;
+
+static int rswitch_xen_probe(struct platform_device *pdev)
+{
+	if (rswitch_dev)
+		return -EBUSY;
+
+	rswitch_regs = devm_platform_get_and_ioremap_resource(pdev, 0, NULL);
+	if (IS_ERR_OR_NULL(rswitch_regs)) {
+		rswitch_regs = NULL;
+		return -EINVAL;
+	}
+
+	rswitch_irq = platform_get_irq(pdev, 0);
+	if (rswitch_irq < 0)
+		return rswitch_irq;
+
+	rswitch_dev = &pdev->dev;
+	return 0;
+}
+
+static int rswitch_xen_remove(struct platform_device *pdev)
+{
+	/* FIXME: support this */
+
+	rswitch_dev = NULL;
+	rswitch_regs = NULL;
+	return 0;
+}
+
+static const struct of_device_id rswitch_xen_of_table[] = {
+	{ .compatible = "renesas,etherswitch-xen" },
+	{}
+};
+
+static struct platform_driver rswitch_xen_platform_driver = {
+	.probe = rswitch_xen_probe,
+	.remove = rswitch_xen_remove,
+	.driver = {
+		.name = "rswitch-xen",
+		.of_match_table = rswitch_xen_of_table,
+	}
+};
+
+static int rvp_xenfront_call_xenback(struct rvp_xenfront *rvpx, int command)
+{
+	int ret;
+
+	rvpx->shm->fe_to_be_command = command;
+	notify_remote_via_evtchn(rvpx->shm->fe_to_be_chn);
+
+	ret = wait_event_interruptible(rvpx->wait,
+		READ_ONCE(rvpx->shm->fe_to_be_command) == BE_COMMAND_DONE);
+
+	if (ret == 0) {
+		rmb();
+		ret = rvpx->shm->fe_to_be_result;
+	}
+
+	return ret;
+}
+
+static irqreturn_t rvp_xenfront_comm_interrupt(int irq, void *data)
+{
+	struct rvp_xenfront *rvpx = data;
+
+	if (rvpx->shm->fe_to_be_command)
+		wake_up_interruptible(&rvpx->wait);
+
+	xen_irq_lateeoi(irq, 0);
+	return IRQ_HANDLED;
+}
+
+static int rvp_xenfront_register(struct rswitch_vport_desc *rvpd)
+{
+	struct rvp_xenfront *rvpx = rvpd->fe_drv_priv;
+
+	return rvp_xenfront_call_xenback(rvpx, BE_COMMAND_REGISTER);
+}
+
+static void rvp_xenfront_unregister(struct rswitch_vport_desc *rvpd)
+{
+	struct rvp_xenfront *rvpx = rvpd->fe_drv_priv;
+
+	rvp_xenfront_call_xenback(rvpx, BE_COMMAND_UNREGISTER);
+}
+
+static int rvp_xenfront_attach_frontend_rings(struct rswitch_vport_desc *rvpd)
+{
+	struct rvp_xenfront *rvpx = rvpd->fe_drv_priv;
+
+	return rvp_xenfront_call_xenback(rvpx, BE_COMMAND_ATTACH_RINGS);
+}
+
+static int rvp_xenfront_detach_frontend_rings(struct rswitch_vport_desc *rvpd)
+{
+	struct rvp_xenfront *rvpx = rvpd->fe_drv_priv;
+
+	return rvp_xenfront_call_xenback(rvpx, BE_COMMAND_DETACH_RINGS);
+}
+
+static const struct rswitch_vport_desc_ops rvpx_ops = {
+	.vport_register = rvp_xenfront_register,
+	.vport_unregister = rvp_xenfront_unregister,
+	.attach_frontend_rings = rvp_xenfront_attach_frontend_rings,
+	.detach_frontend_rings = rvp_xenfront_detach_frontend_rings,
+};
+
+static int rvp_xenfront_run(struct rvp_xenfront *rvpx)
+{
+	struct rswitch_vport_desc *rvpd = &rvpx->shm->rvpd;
+	int osid;
+	int ret;
+
+	ret = xenbus_gather(XBT_NIL, rvpx->xdev->nodename,
+			"osid", "%d", &osid, NULL);
+	if (ret < 0) {
+		xenbus_dev_error(rvpx->xdev, ret, "osid read failed");
+		return ret;
+	}
+
+	rvpd->ops = &rvpx_ops;
+	rvpd->id = RSWITCH_VPORT_ID_DYNAMIC;
+	rvpd->osid = osid;
+	rvpd->irq_index = osid;		/* FIXME */
+	rvpd->fe_reg_base = rswitch_regs;
+	rvpd->fe_irq = rswitch_irq;
+	rvpd->dma_dev = &rvpx->xdev->dev;
+	rvpd->fe_drv_priv = rvpx;
+
+	return rswitch_vport_frontend_run(rvpd);
+}
+
+static void rvp_xenfront_stop(struct rvp_xenfront *rvpx)
+{
+	struct rswitch_vport_desc *rvpd = &rvpx->shm->rvpd;
+
+	rswitch_vport_frontend_stop(rvpd);
+}
+
+static void rvp_xenfront_backend_changed(struct xenbus_device *dev,
+		enum xenbus_state backend_state)
+{
+	struct rvp_xenfront *rvpx = dev_get_drvdata(&dev->dev);
+	int ret;
+
+	if (backend_state == XenbusStateConnected && !rvpx->rvpf_running) {
+		ret = rvp_xenfront_run(rvpx);
+		if (ret == 0)
+			rvpx->rvpf_running = true;
+	}
+
+	if (backend_state != XenbusStateConnected && rvpx->rvpf_running) {
+		rvp_xenfront_stop(rvpx);
+		rvpx->rvpf_running = false;
+	}
+}
+
+static int rvp_xenfront_probe(struct xenbus_device *dev,
+		const struct xenbus_device_id *id)
+{
+	struct rvp_xenfront *rvpx;
+	struct gnttab_map_grant_ref map_op;
+	struct gnttab_unmap_grant_ref unmap_op;
+	int ret;
+
+	xenbus_switch_state(dev, XenbusStateInitialising);
+
+	dma_coerce_mask_and_coherent(&dev->dev, DMA_BIT_MASK(40));
+
+	rvpx = kzalloc(sizeof(*rvpx), GFP_KERNEL);
+	if (!rvpx) {
+		ret = -ENOMEM;
+		xenbus_dev_fatal(dev, ret, "alloc etherdev failed");
+		goto alloc_failed;
+	}
+
+	rvpx->xdev = dev;
+	init_waitqueue_head(&rvpx->wait);
+	dev_set_drvdata(&dev->dev, rvpx);
+
+	ret = xenbus_gather(XBT_NIL, dev->otherend,
+			"shm-gref", "%u", &rvpx->shm_gref, NULL);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "shm gref read failed");
+		goto shm_gref_read_failed;
+	}
+
+	ret = gnttab_alloc_pages(1, &rvpx->shm_page);
+	if (ret) {
+		xenbus_dev_fatal(dev, ret, "shm gnttab alloc failed");
+		goto shm_gnttab_alloc_failed;
+	}
+
+	gnttab_set_map_op(&map_op,
+		(uintptr_t)pfn_to_kaddr(page_to_xen_pfn(rvpx->shm_page)),
+		GNTMAP_host_map, rvpx->shm_gref, dev->otherend_id);
+	ret = gnttab_map_refs(&map_op, NULL, &rvpx->shm_page, 1);
+	if (ret) {
+		xenbus_dev_fatal(dev, ret, "shm map failed");
+		goto shm_map_failed;
+	}
+	rvpx->shm = (struct rswitch_vport_xen_shm *)
+		page_address(rvpx->shm_page);
+
+	ret = bind_interdomain_evtchn_to_irqhandler_lateeoi(dev->otherend_id,
+			rvpx->shm->be_to_fe_chn,
+			rvp_xenfront_comm_interrupt, 0, "rvp_xen_comm", rvpx);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "bind comm handler failed");
+		goto bind_comm_handler_failed;
+	}
+	rvpx->from_be_irq = ret;
+
+	ret = xenbus_alloc_evtchn(dev, &rvpx->shm->fe_to_be_chn);
+	if (ret < 0) {
+		xenbus_dev_fatal(dev, ret, "fe_to_be channel alloc failed");
+		goto channel_alloc_failed;
+	}
+
+	xenbus_switch_state(dev, XenbusStateConnected);
+	return 0;
+
+channel_alloc_failed:
+	unbind_from_irqhandler(rvpx->from_be_irq, rvpx);
+bind_comm_handler_failed:
+	gnttab_set_unmap_op(&unmap_op,
+		(uintptr_t)pfn_to_kaddr(page_to_xen_pfn(rvpx->shm_page)),
+		GNTMAP_host_map, rvpx->shm_gref);
+	gnttab_unmap_refs(&unmap_op, NULL, &rvpx->shm_page, 1);
+shm_map_failed:
+	gnttab_free_pages(1, &rvpx->shm_page);
+shm_gnttab_alloc_failed:
+shm_gref_read_failed:
+	kfree(rvpx);
+alloc_failed:
+	xenbus_switch_state(dev, XenbusStateClosed);
+	return ret;
+}
+
+static int rvp_xenfront_remove(struct xenbus_device *dev)
+{
+	struct rvp_xenfront *rvpx = dev_get_drvdata(&dev->dev);
+	struct gnttab_unmap_grant_ref unmap_op;
+
+	if (rvpx->rvpf_running) {
+		rvp_xenfront_stop(rvpx);
+		rvpx->rvpf_running = false;
+	}
+
+	xenbus_switch_state(dev, XenbusStateClosing);
+
+	xenbus_free_evtchn(dev, rvpx->shm->fe_to_be_chn);
+	unbind_from_irqhandler(rvpx->from_be_irq, rvpx);
+	gnttab_set_unmap_op(&unmap_op,
+		(uintptr_t)pfn_to_kaddr(page_to_xen_pfn(rvpx->shm_page)),
+		GNTMAP_host_map, rvpx->shm_gref);
+	gnttab_unmap_refs(&unmap_op, NULL, &rvpx->shm_page, 1);
+	gnttab_free_pages(1, &rvpx->shm_page);
+
+	kfree(rvpx);
+
+	xenbus_switch_state(dev, XenbusStateClosed);
+
+	return 0;
+}
+
+static const struct xenbus_device_id rvp_xenfront_ids[] = {
+	{ "rswitch_vport" },
+	{ "" }
+};
+
+static struct xenbus_driver rvp_xenfront_driver = {
+	.ids = rvp_xenfront_ids,
+	.probe = rvp_xenfront_probe,
+	.remove = rvp_xenfront_remove,
+	.otherend_changed = rvp_xenfront_backend_changed,
+};
+
+static int __init rvp_xenfront_init(void)
+{
+	int ret;
+
+	if (!xen_domain())
+		return -ENODEV;
+
+	ret = platform_driver_register(&rswitch_xen_platform_driver);
+	if (ret)
+		return ret;
+
+	ret = xenbus_register_frontend(&rvp_xenfront_driver);
+	if (ret) {
+		platform_driver_unregister(&rswitch_xen_platform_driver);
+		return ret;
+	}
+
+	return 0;
+}
+module_init(rvp_xenfront_init);
+
+static void __exit rvp_xenfront_exit(void)
+{
+	xenbus_unregister_driver(&rvp_xenfront_driver);
+	platform_driver_unregister(&rswitch_xen_platform_driver);
+}
+module_exit(rvp_xenfront_exit);
+
+MODULE_LICENSE("GPL");
-- 
2.25.1

